<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Mary Piper, Jihe Liu, Meeta Mistry, Radhika Khetani, Will Gammerdinger">

<title>Introduction to R - Tidyverse data wrangling</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="../css/styles.css">
</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Introduction to R</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../lessons/schedule.html" rel="" target="">
 <span class="menu-text">Schedule</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://bioinformatics.sph.harvard.edu/" rel="" target="">
 <span class="menu-text">HBC</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/hbctraining/intro_r_rmd" rel="" target="">
 <span class="menu-text">GitHub</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="mailto:hbctraining@hsph.harvard.edu" rel="" target="">
 <span class="menu-text">Contact us</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../lessons/11b_Custom_Functions_ggplot2.html">Day 3 Self-learning</a></li><li class="breadcrumb-item"><a href="../lessons/15_tidyverse.html">Tidyverse data wrangling</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false">
 <span class="menu-text">Day 1:</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/01b_introR-R-and-RStudio.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction to R and RStudio</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false">
 <span class="menu-text">Day 1 Self-learning:</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/02_introR-syntax-and-data-structures.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">R Syntax and Data Structures</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/03_introR-functions-and-arguments.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Functions in R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/06_reading_and_data_inspection.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Reading in and inspecting data</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false">
 <span class="menu-text">Day 2</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/Day2_activities.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Day 2 Activities</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/05_introR-data-wrangling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Data subsetting with base R: vectors and factors</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false">
 <span class="menu-text">Day 2 Self-learning</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/04_introR_packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Packages and libraries</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/07_introR-data-wrangling2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Data wrangling: dataframes, matrices, and lists</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/08_identifying-matching-elements.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Advanced R, logical operators for matching</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/09_reordering-to-match-datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Advanced R, reordering to match datasets</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/10_setting_up_to_plot.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Plotting and data visualization in R</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">
 <span class="menu-text">Day 3</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/Day3_activities.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Day 3 Activities</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/11_ggplot2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Plotting and data visualization in R</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">
 <span class="menu-text">Day 3 Self-learning</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/11b_Custom_Functions_ggplot2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Custom functions for consistent plots</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/12_boxplot_exercise.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Plotting and data visualization in R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/13_exporting_data_and_plots.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Saving data and plots to file</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/14_finding_help.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Troubleshooting and finding help</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/15_tidyverse.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Tidyverse data wrangling</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="false">
 <span class="menu-text">Day 4</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/Day4_activities.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Day 4 Activities</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link active" data-scroll-target="#learning-objectives">Learning Objectives</a></li>
  <li><a href="#data-wrangling-with-tidyverse" id="toc-data-wrangling-with-tidyverse" class="nav-link" data-scroll-target="#data-wrangling-with-tidyverse">Data Wrangling with Tidyverse</a>
  <ul class="collapse">
  <li><a href="#tidyverse-basics" id="toc-tidyverse-basics" class="nav-link" data-scroll-target="#tidyverse-basics">Tidyverse basics</a>
  <ul class="collapse">
  <li><a href="#pipes" id="toc-pipes" class="nav-link" data-scroll-target="#pipes">Pipes</a></li>
  <li><a href="#tibbles" id="toc-tibbles" class="nav-link" data-scroll-target="#tibbles">Tibbles</a></li>
  </ul></li>
  <li><a href="#experimental-data" id="toc-experimental-data" class="nav-link" data-scroll-target="#experimental-data">Experimental data</a></li>
  <li><a href="#analysis-goal-and-workflow" id="toc-analysis-goal-and-workflow" class="nav-link" data-scroll-target="#analysis-goal-and-workflow">Analysis goal and workflow</a></li>
  <li><a href="#tidyverse-tools" id="toc-tidyverse-tools" class="nav-link" data-scroll-target="#tidyverse-tools">Tidyverse tools</a></li>
  <li><a href="#read-in-the-functional-analysis-results" id="toc-read-in-the-functional-analysis-results" class="nav-link" data-scroll-target="#read-in-the-functional-analysis-results">1. Read in the functional analysis results</a></li>
  <li><a href="#extract-only-the-go-biological-processes-bp-of-interest" id="toc-extract-only-the-go-biological-processes-bp-of-interest" class="nav-link" data-scroll-target="#extract-only-the-go-biological-processes-bp-of-interest">2. Extract only the GO biological processes (BP) of interest</a></li>
  <li><a href="#select-only-the-columns-needed-for-visualization" id="toc-select-only-the-columns-needed-for-visualization" class="nav-link" data-scroll-target="#select-only-the-columns-needed-for-visualization">3. Select only the columns needed for visualization</a></li>
  <li><a href="#order-go-processes-by-significance-adjusted-p-values" id="toc-order-go-processes-by-significance-adjusted-p-values" class="nav-link" data-scroll-target="#order-go-processes-by-significance-adjusted-p-values">4. Order GO processes by significance (adjusted p-values)</a></li>
  <li><a href="#rename-columns-to-be-more-intuitive" id="toc-rename-columns-to-be-more-intuitive" class="nav-link" data-scroll-target="#rename-columns-to-be-more-intuitive">5. Rename columns to be more intuitive</a></li>
  <li><a href="#create-additional-metrics-for-plotting-e.g.-gene-ratios" id="toc-create-additional-metrics-for-plotting-e.g.-gene-ratios" class="nav-link" data-scroll-target="#create-additional-metrics-for-plotting-e.g.-gene-ratios">6. Create additional metrics for plotting (e.g.&nbsp;gene ratios)</a></li>
  <li><a href="#next-steps" id="toc-next-steps" class="nav-link" data-scroll-target="#next-steps">Next steps</a>
  <ul class="collapse">
  <li><a href="#additional-resources" id="toc-additional-resources" class="nav-link" data-scroll-target="#additional-resources">Additional resources</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Tidyverse data wrangling</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Mary Piper, Jihe Liu, Meeta Mistry, Radhika Khetani, Will Gammerdinger </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<p>Approximate time: 75 minutes</p>
<section id="learning-objectives" class="level2">
<h2 class="anchored" data-anchor-id="learning-objectives">Learning Objectives</h2>
<ul>
<li>Perform basic data wrangling with functions in the Tidyverse package.</li>
</ul>
</section>
<section id="data-wrangling-with-tidyverse" class="level1">
<h1>Data Wrangling with Tidyverse</h1>
<p>The <a href="https://www.tidyverse.org/packages/">Tidyverse suite of integrated packages</a> are designed to work together to make common data science operations more user friendly. The packages have functions for data wrangling, tidying, reading/writing, parsing, and visualizing, among others. There is a freely available book, <a href="http://r4ds.had.co.nz/index.html">R for Data Science</a>, with detailed descriptions and practical examples of the tools available and how they work together. We will explore the basic syntax for working with these packages, as well as, specific functions for data wrangling with the ‘dplyr’ package and data visualization with the ‘ggplot2’ package.</p>
<p><img src="../img/tidyverse_website.png" class="img-fluid"></p>
<section id="tidyverse-basics" class="level2">
<h2 class="anchored" data-anchor-id="tidyverse-basics">Tidyverse basics</h2>
<p>The Tidyverse suite of packages introduces users to a set of data structures, functions and operators to make working with data more intuitive, but is slightly different from the way we do things in base R. <strong>Two important new concepts we will focus on are pipes and tibbles</strong>.</p>
<p>Before we get started with pipes or tibbles, let’s load the library:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="pipes" class="level3">
<h3 class="anchored" data-anchor-id="pipes">Pipes</h3>
<p>Stringing together commands in R can be quite daunting. Also, trying to understand code that has many nested functions can be confusing.</p>
<p>To make R code more human readable, the Tidyverse tools use the pipe, <code>%&gt;%</code>, which was acquired from the <code>magrittr</code> package and is now part of the <code>dplyr</code> package that is installed automatically with Tidyverse. <strong>The pipe allows the output of a previous command to be used as input to another command instead of using nested functions.</strong></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Shortcut to write the pipe is <kbd>shift</kbd> + <kbd>command</kbd> + <kbd>M</kbd></p>
</div>
</div>
<p>An example of using the pipe to run multiple commands:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="do">## A single command</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sqrt</span>(<span class="dv">83</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 9.110434</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Base R method of running more than one command</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="fu">sqrt</span>(<span class="dv">83</span>), <span class="at">digits =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 9.11</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Running more than one command with piping</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sqrt</span>(<span class="dv">83</span>) <span class="sc">%&gt;%</span> <span class="fu">round</span>(<span class="at">digits =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 9.11</code></pre>
</div>
</div>
<p>The pipe represents a much easier way of writing and deciphering R code, and so we will be taking advantage of it, when possible, as we work through the remaining lesson.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercises
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li>Create a vector of random numbers using the code below:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>random_numbers <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">81</span>, <span class="dv">90</span>, <span class="dv">65</span>, <span class="dv">43</span>, <span class="dv">71</span>, <span class="dv">29</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="1">
<li><p>Use the pipe (<code>%&gt;%</code>) to perform two steps in a single line:</p>
<ol type="1">
<li>Take the mean of <code>random_numbers</code> using the <code>mean()</code> function.</li>
<li>Round the output to three digits using the <code>round()</code> function.</li>
</ol></li>
</ol>
</div>
</div>
</section>
<section id="tibbles" class="level3">
<h3 class="anchored" data-anchor-id="tibbles">Tibbles</h3>
<p>A core component of the <a href="http://tidyverse.org/">tidyverse</a> is the <a href="http://tibble.tidyverse.org/">tibble</a>. <strong>Tibbles are a modern rework of the standard <code>data.frame</code>, with some internal improvements</strong> to make code more reliable. They are data frames, but do not follow all of the same rules. For example, tibbles can have numbers/symbols for column names, which is not normally allowed in base R.</p>
<p><strong>Important: <a href="http://tidyverse.org/">tidyverse</a> is very opinionated about row names</strong>. These packages insist that all column data (e.g.&nbsp;<code>data.frame</code>) be treated equally, and that special designation of a column as <code>rownames</code> should be deprecated. <a href="http://tibble.tidyverse.org/">Tibble</a> provides simple utility functions to handle rownames: <code>rownames_to_column()</code> and <code>column_to_rownames()</code>.</p>
<p>Tibbles can be created directly using the <code>tibble()</code> function or data frames can be converted into tibbles using <code>as_tibble(name_of_df)</code>.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The function <code>as_tibble()</code> will ignore row names, so if a column representing the row names is needed, then the function <code>rownames_to_column(name_of_df)</code> should be run prior to turning the data.frame into a tibble. Also, <code>as_tibble()</code> will not coerce character vectors to factors by default.</p>
</div>
</div>
</section>
</section>
<section id="experimental-data" class="level2">
<h2 class="anchored" data-anchor-id="experimental-data">Experimental data</h2>
<p>We’re going to explore the Tidyverse suite of tools to wrangle our data to prepare it for visualization. You should have downloaded the file called <code>gprofiler_results_Mov10oe.tsv</code> into your R project’s <code>data</code> folder earlier.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you do not have the <code>gprofiler_results_Mov10oe.tsv</code> file in your <code>data</code> folder, you can right click and download it into the <code>data</code> folder <a href="https://github.com/hbctraining/Training-modules/blob/master/Tidyverse_ggplot2/data/gprofiler_results_Mov10oe.tsv?raw=true">using this link</a>.</p>
</div>
</div>
<p><strong>The dataset:</strong></p>
<ul>
<li>Represents the <strong>functional analysis results</strong>, including the biological processes, functions, pathways, or conditions that are over-represented in a given list of genes.</li>
<li>Our gene list was generated by <strong>differential gene expression analysis</strong> and the genes represent differences between <strong>control mice</strong> and <strong>mice over-expressing a gene involved in RNA splicing</strong>.</li>
</ul>
<p>The functional analysis that we will focus on involves <strong>gene ontology (GO) terms</strong>, which:</p>
<ul>
<li>describe the roles of genes and gene products</li>
<li>organized into three controlled vocabularies/ontologies (domains):
<ul>
<li>biological processes (BP)</li>
<li>cellular components (CC)</li>
<li>molecular functions (MF)</li>
</ul></li>
</ul>
<p><img src="../img/mov10_FA.png" width="800"></p>
</section>
<section id="analysis-goal-and-workflow" class="level2">
<h2 class="anchored" data-anchor-id="analysis-goal-and-workflow">Analysis goal and workflow</h2>
<p><strong>Goal:</strong> <em>Visually compare the most significant biological processes (BP) based on the number of associated differentially expressed genes (gene ratios) and significance values by creating the following plot:</em></p>
<p><img src="../img/dotplot6.png" class="img-fluid"></p>
<p>To wrangle our data in preparation for the plotting, we are going to use the Tidyverse suite of tools to wrangle and visualize our data through several steps:</p>
<ol type="1">
<li>Read in the functional analysis results</li>
<li>Extract only the GO biological processes (BP) of interest</li>
<li>Select only the columns needed for visualization</li>
<li>Order by significance (p-adjusted values)</li>
<li>Rename columns to be more intuitive</li>
<li>Create additional metrics for plotting (e.g.&nbsp;gene ratios)</li>
<li>Plot results</li>
</ol>
</section>
<section id="tidyverse-tools" class="level2">
<h2 class="anchored" data-anchor-id="tidyverse-tools">Tidyverse tools</h2>
<p>While all of the tools in the Tidyverse suite are deserving of being explored in more depth, we are going to investigate more deeply the reading (<code>readr</code>), wrangling (<code>dplyr</code>), and plotting (<code>ggplot2</code>) tools.</p>
</section>
<section id="read-in-the-functional-analysis-results" class="level2">
<h2 class="anchored" data-anchor-id="read-in-the-functional-analysis-results">1. Read in the functional analysis results</h2>
<p>While the base R packages have perfectly fine methods for reading in data, the <code>readr</code> and <code>readxl</code> Tidyverse packages offer additional methods for reading in data. Let’s read in our tab-delimited functional analysis results using <code>read_delim()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the functional analysis results</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>functional_GO_results <span class="ot">&lt;-</span> <span class="fu">read_delim</span>(<span class="at">file =</span> <span class="st">"data/gprofiler_results_Mov10oe.txt"</span>, <span class="at">delim =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span> )</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look at the results</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>functional_GO_results</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3,644 × 14
   query.number significant p.value term.size query.size overlap.size recall
          &lt;dbl&gt; &lt;lgl&gt;         &lt;dbl&gt;     &lt;dbl&gt;      &lt;dbl&gt;        &lt;dbl&gt;  &lt;dbl&gt;
 1            1 TRUE        0.00434       111       5850           52  0.009
 2            1 TRUE        0.0033        110       5850           52  0.009
 3            1 TRUE        0.0297         39       5850           21  0.004
 4            1 TRUE        0.0193         70       5850           34  0.006
 5            1 TRUE        0.0148         26       5850           16  0.003
 6            1 TRUE        0.0187         22       5850           14  0.002
 7            1 TRUE        0.0226         48       5850           25  0.004
 8            1 TRUE        0.0491         17       5850           11  0.002
 9            1 TRUE        0.00798       164       5850           71  0.012
10            1 TRUE        0.0439         19       5850           12  0.002
# ℹ 3,634 more rows
# ℹ 7 more variables: precision &lt;dbl&gt;, term.id &lt;chr&gt;, domain &lt;chr&gt;,
#   subgraph.number &lt;dbl&gt;, term.name &lt;chr&gt;, relative.depth &lt;dbl&gt;,
#   intersection &lt;chr&gt;</code></pre>
</div>
</div>
<details>
<summary>
<i>Click here to see how to do this in base R</i>
</summary>
<p><br></p>
<pre><code># Read in the functional analysis results<br>
functional_GO_results &lt;- read.delim(file = "data/gprofiler_results_Mov10oe.tsv", sep = "\t" )</code><br>
<code># Take a look at the results
functional_GO_results</code></pre>
<p><br></p>
</details>
<p>Notice that the results were automatically read in as a tibble and the output gives the number of rows, columns and the data type for each of the columns.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>A large number of tidyverse functions will work with both tibbles and dataframes, and the data structure of the output will be identical to the input. However, there are some functions that will return a tibble (without row names), whether or not a tibble or dataframe is provided.</p>
</div>
</div>
</section>
<section id="extract-only-the-go-biological-processes-bp-of-interest" class="level2">
<h2 class="anchored" data-anchor-id="extract-only-the-go-biological-processes-bp-of-interest">2. Extract only the GO biological processes (BP) of interest</h2>
<p>Now that we have our data, we will need to wrangle it into a format ready for plotting. For all of our data wrangling steps we will be using tools from the <a href="http://dplyr.tidyverse.org/">dplyr</a> package, which is a swiss-army knife for data wrangling of data frames.</p>
<p>To extract the biological processes of interest, we only want those rows where the <code>domain</code> is equal to <code>BP</code>, which we can do using the <code>filter()</code> function.</p>
<p>To filter rows of a data frame/tibble based on values in different columns, we give a logical expression as input to the <code>filter()</code> function to return those rows for which the expression is TRUE.</p>
<p>Now let’s return only those rows that have a <code>domain</code> of <code>BP</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Return only GO biological processes</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>bp_oe <span class="ot">&lt;-</span> functional_GO_results <span class="sc">%&gt;%</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(domain <span class="sc">==</span> <span class="st">"BP"</span>)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>bp_oe</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,102 × 14
   query.number significant p.value term.size query.size overlap.size recall
          &lt;dbl&gt; &lt;lgl&gt;         &lt;dbl&gt;     &lt;dbl&gt;      &lt;dbl&gt;        &lt;dbl&gt;  &lt;dbl&gt;
 1            1 TRUE        0.00434       111       5850           52  0.009
 2            1 TRUE        0.0033        110       5850           52  0.009
 3            1 TRUE        0.0297         39       5850           21  0.004
 4            1 TRUE        0.0193         70       5850           34  0.006
 5            1 TRUE        0.0148         26       5850           16  0.003
 6            1 TRUE        0.0187         22       5850           14  0.002
 7            1 TRUE        0.0226         48       5850           25  0.004
 8            1 TRUE        0.0491         17       5850           11  0.002
 9            1 TRUE        0.00798       164       5850           71  0.012
10            1 TRUE        0.0439         19       5850           12  0.002
# ℹ 1,092 more rows
# ℹ 7 more variables: precision &lt;dbl&gt;, term.id &lt;chr&gt;, domain &lt;chr&gt;,
#   subgraph.number &lt;dbl&gt;, term.name &lt;chr&gt;, relative.depth &lt;dbl&gt;,
#   intersection &lt;chr&gt;</code></pre>
</div>
</div>
<details>
<summary>
<i>Click here to see how to do this in base R</i>
</summary>
<p><br></p>
<pre><code># Return only GO biological processes
idx &lt;- functional_GO_results$domain == "BP"
bp_oe2 &lt;- functional_GO_results[idx,]</code><br>
<code>View(bp_oe)</code></pre>
<p><br></p>
</details>
<p>Now we have returned only those rows with a <code>domain</code> of <code>BP</code>. <strong>How have the dimensions of our results changed?</strong></p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>We would like to perform an additional round of filtering to only keep the most specific GO terms.</p>
<ol type="1">
<li>For <code>bp_oe</code>, use the <code>filter()</code> function to only keep those rows where the <code>relative.depth</code> is greater than 4.</li>
<li>Save output to overwrite our <code>bp_oe</code> variable.</li>
</ol>
</div>
</div>
</section>
<section id="select-only-the-columns-needed-for-visualization" class="level2">
<h2 class="anchored" data-anchor-id="select-only-the-columns-needed-for-visualization">3. Select only the columns needed for visualization</h2>
<p>For visualization purposes, we are only interested in the columns related to the GO terms, the significance of the terms, and information about the number of genes associated with the terms.</p>
<p>To extract columns from a data frame/tibble we can use the <code>select()</code> function. In contrast to base R, we do not need to put the column names in quotes for selection.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Selecting columns to keep</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>bp_oe <span class="ot">&lt;-</span> bp_oe <span class="sc">%&gt;%</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(term.id, term.name, p.value, query.size, term.size, overlap.size, intersection)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>bp_oe</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,102 × 7
   term.id    term.name   p.value query.size term.size overlap.size intersection
   &lt;chr&gt;      &lt;chr&gt;         &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;        &lt;dbl&gt; &lt;chr&gt;       
 1 GO:0032606 type I int… 0.00434       5850       111           52 rnf216,polr…
 2 GO:0032479 regulation… 0.0033        5850       110           52 rnf216,polr…
 3 GO:0032480 negative r… 0.0297        5850        39           21 rnf216,otud…
 4 GO:0032481 positive r… 0.0193        5850        70           34 polr3b,polr…
 5 GO:0010644 cell commu… 0.0148        5850        26           16 pkp2,prkaca…
 6 GO:0086064 cell commu… 0.0187        5850        22           14 pkp2,prkaca…
 7 GO:0035904 aorta deve… 0.0226        5850        48           25 plxnd1,ap2b…
 8 GO:0034199 activation… 0.0491        5850        17           11 prkar2b,prk…
 9 GO:0050852 T cell rec… 0.00798       5850       164           71 psmb1,psmc4…
10 GO:0046653 tetrahydro… 0.0439        5850        19           12 mthfd1,mthf…
# ℹ 1,092 more rows</code></pre>
</div>
</div>
<details>
<summary>
<i>Click here to see how to do this in base R</i>
</summary>
<p><br></p>
<pre><code># Selecting columns to keep
bp_oe &lt;- bp_oe[, c("term.id", "term.name", "p.value", "query.size", "term.size", "overlap.size", "intersection")]</code><br>
<code>View(bp_oe)</code></pre>
<p><br></p>
</details>
<p>The <code>select()</code> function also allows for negative selection. So we could have alternately removed columns with negative selection. Note that we need to put the column names inside of the combine (<code>c()</code>) function with a <code>-</code> preceding it for this functionality.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Selecting columns to remove</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>bp_oe <span class="ot">&lt;-</span> bp_oe <span class="sc">%&gt;%</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(query.number, significant, recall, precision, </span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>              subgraph.number, relative.depth, domain))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<details>
<summary>
<i>Click here to see how to do this in base R</i>
</summary>
<p><br></p>
<pre><code># DO NOT RUN</code>
<code># Selecting columns to remove
idx &lt;- !(colnames(bp_oe) %in% c("query.number", "significant", "recall", "precision", "subgraph.number", "relative.depth", "domain"))
bp_oe &lt;- bp_oe[, idx]</code></pre>
<p><br></p>
</details>
<p><img src="../img/bp_oe_selection.png" width="800"></p>
</section>
<section id="order-go-processes-by-significance-adjusted-p-values" class="level2">
<h2 class="anchored" data-anchor-id="order-go-processes-by-significance-adjusted-p-values">4. Order GO processes by significance (adjusted p-values)</h2>
<p>Now that we have only the rows and columns of interest, let’s arrange these by significance, which is denoted by the adjusted p-value.</p>
<p>Let’s sort the rows by adjusted p-value with the <code>arrange()</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Order by adjusted p-value ascending</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>bp_oe <span class="ot">&lt;-</span> bp_oe <span class="sc">%&gt;%</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(p.value)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<details>
<summary>
<i>Click here to see how to do this in base R</i>
</summary>
<p><br></p>
<pre><code># Order by adjusted p-value ascending
idx &lt;- order(bp_oe$p.value)
bp_oe &lt;- bp_oe[idx,]</code></pre>
<p><br></p>
</details>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you wanted to arrange in descending order, then you could have run the following instead:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Order by adjusted p-value descending</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>bp_oe <span class="ot">&lt;-</span> bp_oe <span class="sc">%&gt;%</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(p.value))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<details>
<summary>
<i>Click here to see how to do this in base R</i>
</summary>
<p><br></p>
<pre><code># DO NOT RUN</code>
<code># Order by adjusted p-value descending
idx &lt;- order(bp_oe$p.value, decreasing = TRUE)
bp_oe &lt;- bp_oe[idx,]</code></pre>
</details>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Ordering variables in <code>ggplot2</code> is a bit different. <a href="https://www.r-graph-gallery.com/267-reorder-a-variable-in-ggplot2.html">This post</a> introduces a few ways of ordering variables in a plot.</p>
</div>
</div>
</section>
<section id="rename-columns-to-be-more-intuitive" class="level2">
<h2 class="anchored" data-anchor-id="rename-columns-to-be-more-intuitive">5. Rename columns to be more intuitive</h2>
<p>While not necessary for our visualization, renaming columns more intuitively can help with our understanding of the data using the <code>rename()</code> function. The syntax is <code>new_name</code> = <code>old_name</code>.</p>
<p>Let’s rename the <code>term.id</code> and <code>term.name</code> columns.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Provide better names for columns</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>bp_oe <span class="ot">&lt;-</span> bp_oe <span class="sc">%&gt;%</span> </span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">GO_id =</span> term.id, </span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">GO_term =</span> term.name)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(bp_oe)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 7
  GO_id      GO_term      p.value query.size term.size overlap.size intersection
  &lt;chr&gt;      &lt;chr&gt;          &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;        &lt;dbl&gt; &lt;chr&gt;       
1 GO:0044260 cellular m… 1.56e-75       5850      8276         3171 dpm1,scyl3,…
2 GO:0034641 cellular n… 3.51e-66       5850      6428         2535 dpm1,gclc,n…
3 GO:0010467 gene expre… 6.71e-66       5850      5257         2142 gclc,nfya,b…
4 GO:0006807 nitrogen c… 2.85e-64       5850      6866         2671 dpm1,gclc,n…
5 GO:0044237 cellular m… 3.27e-64       5850     10105         3695 dpm1,scyl3,…
6 GO:0090304 nucleic ac… 1.18e-61       5850      5103         2073 gclc,nfya,d…</code></pre>
</div>
</div>
<details>
<summary>
<i>Click here to see how to do this in base R</i>
</summary>
<p><br></p>
<pre><code># Provide better names for columns
colnames(bp_oe)[colnames(bp_oe) == "term.id"] &lt;- "GO_id"
colnames(bp_oe)[colnames(bp_oe) == "term.name"] &lt;- "GO_term"</code></pre>
<p><br></p>
</details>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>In the case of two packages with identical function names, you can use <code>::</code> with the package name before and the function name after (e.g <code>stats::filter()</code>) to ensure that the correct function is implemented. The <code>::</code> can also be used to bring in a function from a library without loading it first.</p>
<p>In the example above, we wanted to use the <code>rename()</code> function specifically from the <code>dplyr</code> package, and not any of the other packages (or base R) which may have the <code>rename()</code> function.</p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Rename the <code>intersection</code> column to <code>genes</code> to reflect the fact that these are the DE genes associated with the GO process.</p>
</div>
</div>
</section>
<section id="create-additional-metrics-for-plotting-e.g.-gene-ratios" class="level2">
<h2 class="anchored" data-anchor-id="create-additional-metrics-for-plotting-e.g.-gene-ratios">6. Create additional metrics for plotting (e.g.&nbsp;gene ratios)</h2>
<p>Finally, before we plot our data, we need to create a couple of additional metrics. The <code>mutate()</code> function enables you to create a new column from an existing column.</p>
<p>Let’s generate gene ratios to reflect the number of DE genes associated with each GO process relative to the total number of DE genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create gene ratio column based on other columns in dataset</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>bp_oe <span class="ot">&lt;-</span> bp_oe <span class="sc">%&gt;%</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">gene_ratio =</span> overlap.size <span class="sc">/</span> query.size)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(bp_oe)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
  GO_id      GO_term      p.value query.size term.size overlap.size intersection
  &lt;chr&gt;      &lt;chr&gt;          &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;        &lt;dbl&gt; &lt;chr&gt;       
1 GO:0044260 cellular m… 1.56e-75       5850      8276         3171 dpm1,scyl3,…
2 GO:0034641 cellular n… 3.51e-66       5850      6428         2535 dpm1,gclc,n…
3 GO:0010467 gene expre… 6.71e-66       5850      5257         2142 gclc,nfya,b…
4 GO:0006807 nitrogen c… 2.85e-64       5850      6866         2671 dpm1,gclc,n…
5 GO:0044237 cellular m… 3.27e-64       5850     10105         3695 dpm1,scyl3,…
6 GO:0090304 nucleic ac… 1.18e-61       5850      5103         2073 gclc,nfya,d…
# ℹ 1 more variable: gene_ratio &lt;dbl&gt;</code></pre>
</div>
</div>
<details>
<summary>
<i>Click here to see how to do this in base R</i>
</summary>
<p><br></p>
<pre><code># Create gene ratio column based on other columns in dataset
bp_oe &lt;- cbind(bp_oe, gene_ratio = bp_oe$overlap.size / bp_oe$query.size)</code></pre>
<p><br></p>
</details>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Create a column in <code>bp_oe</code> called <code>term_percent</code> to determine the percent of DE genes associated with the GO term relative to the total number of genes associated with the GO term (<code>overlap.size</code> / <code>term.size</code>)</p>
</div>
</div>
<p>Our final data for plotting should look like the table below:</p>
<p><img src="../img/bp_oe_ready_to_plot.png" width="900"></p>
</section>
<section id="next-steps" class="level2">
<h2 class="anchored" data-anchor-id="next-steps">Next steps</h2>
<p>Now that we have our results ready for plotting, we can use the <a href="https://ggplot2.tidyverse.org">ggplot2</a> package to plot our results. If you are interested, you can follow <a href="https://hbctraining.github.io/Training-modules/Tidyverse_ggplot2/lessons/03_ggplot2.html">this lesson</a> and dive into how to use <code>ggplot2</code> to create the plots with this dataset.</p>
<section id="additional-resources" class="level3">
<h3 class="anchored" data-anchor-id="additional-resources">Additional resources</h3>
<ul>
<li><a href="http://r4ds.had.co.nz">R for Data Science</a></li>
<li><a href="http://varianceexplained.org/r/teach-tidyverse/">teach the tidyverse</a></li>
<li><a href="http://style.tidyverse.org/">tidy style guide</a></li>
</ul>


<!-- -->

</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb22" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> Tidyverse data wrangling</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> Mary Piper, Jihe Liu, Meeta Mistry, Radhika Khetani, Will Gammerdinger</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>Approximate time: 75 minutes</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a><span class="fu">## Learning Objectives</span></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Perform basic data wrangling with functions in the Tidyverse package.</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a><span class="fu"># Data Wrangling with Tidyverse</span></span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a>The <span class="co">[</span><span class="ot">Tidyverse suite of integrated packages</span><span class="co">](https://www.tidyverse.org/packages/)</span> are designed to work together to make common data science operations more user friendly. The packages have functions for data wrangling, tidying, reading/writing, parsing, and visualizing, among others. There is a freely available book, <span class="co">[</span><span class="ot">R for Data Science</span><span class="co">](http://r4ds.had.co.nz/index.html)</span>, with detailed descriptions and practical examples of the tools available and how they work together. We will explore the basic syntax for working with these packages, as well as, specific functions for data wrangling with the 'dplyr' package and data visualization with the 'ggplot2' package.</span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a><span class="al">![](../img/tidyverse_website.png)</span></span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a><span class="fu">## Tidyverse basics</span></span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a>The Tidyverse suite of packages introduces users to a set of data structures, functions and operators to make working with data more intuitive, but is slightly different from the way we do things in base R. **Two important new concepts we will focus on are pipes and tibbles**.</span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a>Before we get started with pipes or tibbles, let's load the library:</span>
<span id="cb22-23"><a href="#cb22-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-26"><a href="#cb22-26" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb22-27"><a href="#cb22-27" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb22-28"><a href="#cb22-28" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-29"><a href="#cb22-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-30"><a href="#cb22-30" aria-hidden="true" tabindex="-1"></a><span class="fu">### Pipes</span></span>
<span id="cb22-31"><a href="#cb22-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-32"><a href="#cb22-32" aria-hidden="true" tabindex="-1"></a>Stringing together commands in R can be quite daunting. Also, trying to understand code that has many nested functions can be confusing.</span>
<span id="cb22-33"><a href="#cb22-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-34"><a href="#cb22-34" aria-hidden="true" tabindex="-1"></a>To make R code more human readable, the Tidyverse tools use the pipe, <span class="in">`%&gt;%`</span>, which was acquired from the <span class="in">`magrittr`</span> package and is now part of the <span class="in">`dplyr`</span> package that is installed automatically with Tidyverse. **The pipe allows the output of a previous command to be used as input to another command instead of using nested functions.**</span>
<span id="cb22-35"><a href="#cb22-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-36"><a href="#cb22-36" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb22-37"><a href="#cb22-37" aria-hidden="true" tabindex="-1"></a>Shortcut to write the pipe is <span class="kw">&lt;kbd&gt;</span>shift<span class="kw">&lt;/kbd&gt;</span> + <span class="kw">&lt;kbd&gt;</span>command<span class="kw">&lt;/kbd&gt;</span> + <span class="kw">&lt;kbd&gt;</span>M<span class="kw">&lt;/kbd&gt;</span></span>
<span id="cb22-38"><a href="#cb22-38" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb22-39"><a href="#cb22-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-40"><a href="#cb22-40" aria-hidden="true" tabindex="-1"></a>An example of using the pipe to run multiple commands:</span>
<span id="cb22-41"><a href="#cb22-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-44"><a href="#cb22-44" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb22-45"><a href="#cb22-45" aria-hidden="true" tabindex="-1"></a><span class="do">## A single command</span></span>
<span id="cb22-46"><a href="#cb22-46" aria-hidden="true" tabindex="-1"></a><span class="fu">sqrt</span>(<span class="dv">83</span>)</span>
<span id="cb22-47"><a href="#cb22-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-48"><a href="#cb22-48" aria-hidden="true" tabindex="-1"></a><span class="do">## Base R method of running more than one command</span></span>
<span id="cb22-49"><a href="#cb22-49" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="fu">sqrt</span>(<span class="dv">83</span>), <span class="at">digits =</span> <span class="dv">2</span>)</span>
<span id="cb22-50"><a href="#cb22-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-51"><a href="#cb22-51" aria-hidden="true" tabindex="-1"></a><span class="do">## Running more than one command with piping</span></span>
<span id="cb22-52"><a href="#cb22-52" aria-hidden="true" tabindex="-1"></a><span class="fu">sqrt</span>(<span class="dv">83</span>) <span class="sc">%&gt;%</span> <span class="fu">round</span>(<span class="at">digits =</span> <span class="dv">2</span>)</span>
<span id="cb22-53"><a href="#cb22-53" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-54"><a href="#cb22-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-55"><a href="#cb22-55" aria-hidden="true" tabindex="-1"></a>The pipe represents a much easier way of writing and deciphering R code, and so we will be taking advantage of it, when possible, as we work through the remaining lesson.</span>
<span id="cb22-56"><a href="#cb22-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-57"><a href="#cb22-57" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb22-58"><a href="#cb22-58" aria-hidden="true" tabindex="-1"></a><span class="fu"># Exercises</span></span>
<span id="cb22-59"><a href="#cb22-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-60"><a href="#cb22-60" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Create a vector of random numbers using the code below:</span>
<span id="cb22-61"><a href="#cb22-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-64"><a href="#cb22-64" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb22-65"><a href="#cb22-65" aria-hidden="true" tabindex="-1"></a>random_numbers <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">81</span>, <span class="dv">90</span>, <span class="dv">65</span>, <span class="dv">43</span>, <span class="dv">71</span>, <span class="dv">29</span>)</span>
<span id="cb22-66"><a href="#cb22-66" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-67"><a href="#cb22-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-68"><a href="#cb22-68" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Use the pipe (<span class="in">`%&gt;%`</span>) to perform two steps in a single line:</span>
<span id="cb22-69"><a href="#cb22-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-70"><a href="#cb22-70" aria-hidden="true" tabindex="-1"></a><span class="ss">    1.  </span>Take the mean of <span class="in">`random_numbers`</span> using the <span class="in">`mean()`</span> function.</span>
<span id="cb22-71"><a href="#cb22-71" aria-hidden="true" tabindex="-1"></a><span class="ss">    2.  </span>Round the output to three digits using the <span class="in">`round()`</span> function.</span>
<span id="cb22-72"><a href="#cb22-72" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb22-73"><a href="#cb22-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-74"><a href="#cb22-74" aria-hidden="true" tabindex="-1"></a><span class="fu">### Tibbles</span></span>
<span id="cb22-75"><a href="#cb22-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-76"><a href="#cb22-76" aria-hidden="true" tabindex="-1"></a>A core component of the <span class="co">[</span><span class="ot">tidyverse</span><span class="co">](http://tidyverse.org/)</span> is the <span class="co">[</span><span class="ot">tibble</span><span class="co">](http://tibble.tidyverse.org/)</span>. **Tibbles are a modern rework of the standard `data.frame`, with some internal improvements** to make code more reliable. They are data frames, but do not follow all of the same rules. For example, tibbles can have numbers/symbols for column names, which is not normally allowed in base R.</span>
<span id="cb22-77"><a href="#cb22-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-78"><a href="#cb22-78" aria-hidden="true" tabindex="-1"></a>**Important: [tidyverse](http://tidyverse.org/) is very opinionated about row names**. These packages insist that all column data (e.g. <span class="in">`data.frame`</span>) be treated equally, and that special designation of a column as <span class="in">`rownames`</span> should be deprecated. <span class="co">[</span><span class="ot">Tibble</span><span class="co">](http://tibble.tidyverse.org/)</span> provides simple utility functions to handle rownames: <span class="in">`rownames_to_column()`</span> and <span class="in">`column_to_rownames()`</span>.</span>
<span id="cb22-79"><a href="#cb22-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-80"><a href="#cb22-80" aria-hidden="true" tabindex="-1"></a>Tibbles can be created directly using the <span class="in">`tibble()`</span> function or data frames can be converted into tibbles using <span class="in">`as_tibble(name_of_df)`</span>.</span>
<span id="cb22-81"><a href="#cb22-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-82"><a href="#cb22-82" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb22-83"><a href="#cb22-83" aria-hidden="true" tabindex="-1"></a>The function <span class="in">`as_tibble()`</span> will ignore row names, so if a column representing the row names is needed, then the function <span class="in">`rownames_to_column(name_of_df)`</span> should be run prior to turning the data.frame into a tibble. Also, <span class="in">`as_tibble()`</span> will not coerce character vectors to factors by default.</span>
<span id="cb22-84"><a href="#cb22-84" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb22-85"><a href="#cb22-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-86"><a href="#cb22-86" aria-hidden="true" tabindex="-1"></a><span class="fu">## Experimental data</span></span>
<span id="cb22-87"><a href="#cb22-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-88"><a href="#cb22-88" aria-hidden="true" tabindex="-1"></a>We're going to explore the Tidyverse suite of tools to wrangle our data to prepare it for visualization. You should have downloaded the file called <span class="in">`gprofiler_results_Mov10oe.tsv`</span> into your R project's <span class="in">`data`</span> folder earlier.</span>
<span id="cb22-89"><a href="#cb22-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-90"><a href="#cb22-90" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb22-91"><a href="#cb22-91" aria-hidden="true" tabindex="-1"></a>If you do not have the <span class="in">`gprofiler_results_Mov10oe.tsv`</span> file in your <span class="in">`data`</span> folder, you can right click and download it into the <span class="in">`data`</span> folder <span class="co">[</span><span class="ot">using this link</span><span class="co">](https://github.com/hbctraining/Training-modules/blob/master/Tidyverse_ggplot2/data/gprofiler_results_Mov10oe.tsv?raw=true)</span>.</span>
<span id="cb22-92"><a href="#cb22-92" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb22-93"><a href="#cb22-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-94"><a href="#cb22-94" aria-hidden="true" tabindex="-1"></a>**The dataset:**</span>
<span id="cb22-95"><a href="#cb22-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-96"><a href="#cb22-96" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Represents the **functional analysis results**, including the biological processes, functions, pathways, or conditions that are over-represented in a given list of genes.</span>
<span id="cb22-97"><a href="#cb22-97" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Our gene list was generated by **differential gene expression analysis** and the genes represent differences between **control mice** and **mice over-expressing a gene involved in RNA splicing**.</span>
<span id="cb22-98"><a href="#cb22-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-99"><a href="#cb22-99" aria-hidden="true" tabindex="-1"></a>The functional analysis that we will focus on involves **gene ontology (GO) terms**, which:</span>
<span id="cb22-100"><a href="#cb22-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-101"><a href="#cb22-101" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>describe the roles of genes and gene products</span>
<span id="cb22-102"><a href="#cb22-102" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>organized into three controlled vocabularies/ontologies (domains):</span>
<span id="cb22-103"><a href="#cb22-103" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>biological processes (BP)</span>
<span id="cb22-104"><a href="#cb22-104" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>cellular components (CC)</span>
<span id="cb22-105"><a href="#cb22-105" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>molecular functions (MF)</span>
<span id="cb22-106"><a href="#cb22-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-107"><a href="#cb22-107" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;img</span> <span class="er">src</span><span class="ot">=</span><span class="st">"../img/mov10_FA.png"</span> <span class="er">width</span><span class="ot">=</span><span class="st">"800"</span><span class="kw">/&gt;</span></span>
<span id="cb22-108"><a href="#cb22-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-109"><a href="#cb22-109" aria-hidden="true" tabindex="-1"></a><span class="fu">## Analysis goal and workflow</span></span>
<span id="cb22-110"><a href="#cb22-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-111"><a href="#cb22-111" aria-hidden="true" tabindex="-1"></a>**Goal:** *Visually compare the most significant biological processes (BP) based on the number of associated differentially expressed genes (gene ratios) and significance values by creating the following plot:*</span>
<span id="cb22-112"><a href="#cb22-112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-113"><a href="#cb22-113" aria-hidden="true" tabindex="-1"></a><span class="al">![](../img/dotplot6.png)</span></span>
<span id="cb22-114"><a href="#cb22-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-115"><a href="#cb22-115" aria-hidden="true" tabindex="-1"></a>To wrangle our data in preparation for the plotting, we are going to use the Tidyverse suite of tools to wrangle and visualize our data through several steps:</span>
<span id="cb22-116"><a href="#cb22-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-117"><a href="#cb22-117" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Read in the functional analysis results</span>
<span id="cb22-118"><a href="#cb22-118" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Extract only the GO biological processes (BP) of interest</span>
<span id="cb22-119"><a href="#cb22-119" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>Select only the columns needed for visualization</span>
<span id="cb22-120"><a href="#cb22-120" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>Order by significance (p-adjusted values)</span>
<span id="cb22-121"><a href="#cb22-121" aria-hidden="true" tabindex="-1"></a><span class="ss">5.  </span>Rename columns to be more intuitive</span>
<span id="cb22-122"><a href="#cb22-122" aria-hidden="true" tabindex="-1"></a><span class="ss">6.  </span>Create additional metrics for plotting (e.g. gene ratios)</span>
<span id="cb22-123"><a href="#cb22-123" aria-hidden="true" tabindex="-1"></a><span class="ss">7.  </span>Plot results</span>
<span id="cb22-124"><a href="#cb22-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-125"><a href="#cb22-125" aria-hidden="true" tabindex="-1"></a><span class="fu">## Tidyverse tools</span></span>
<span id="cb22-126"><a href="#cb22-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-127"><a href="#cb22-127" aria-hidden="true" tabindex="-1"></a>While all of the tools in the Tidyverse suite are deserving of being explored in more depth, we are going to investigate more deeply the reading (<span class="in">`readr`</span>), wrangling (<span class="in">`dplyr`</span>), and plotting (<span class="in">`ggplot2`</span>) tools.</span>
<span id="cb22-128"><a href="#cb22-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-129"><a href="#cb22-129" aria-hidden="true" tabindex="-1"></a><span class="fu">## 1. Read in the functional analysis results</span></span>
<span id="cb22-130"><a href="#cb22-130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-131"><a href="#cb22-131" aria-hidden="true" tabindex="-1"></a>While the base R packages have perfectly fine methods for reading in data, the <span class="in">`readr`</span> and <span class="in">`readxl`</span> Tidyverse packages offer additional methods for reading in data. Let's read in our tab-delimited functional analysis results using <span class="in">`read_delim()`</span>:</span>
<span id="cb22-132"><a href="#cb22-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-135"><a href="#cb22-135" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb22-136"><a href="#cb22-136" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the functional analysis results</span></span>
<span id="cb22-137"><a href="#cb22-137" aria-hidden="true" tabindex="-1"></a>functional_GO_results <span class="ot">&lt;-</span> <span class="fu">read_delim</span>(<span class="at">file =</span> <span class="st">"data/gprofiler_results_Mov10oe.txt"</span>, <span class="at">delim =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span> )</span>
<span id="cb22-138"><a href="#cb22-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-139"><a href="#cb22-139" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look at the results</span></span>
<span id="cb22-140"><a href="#cb22-140" aria-hidden="true" tabindex="-1"></a>functional_GO_results</span>
<span id="cb22-141"><a href="#cb22-141" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-142"><a href="#cb22-142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-143"><a href="#cb22-143" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;details&gt;</span></span>
<span id="cb22-144"><a href="#cb22-144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-145"><a href="#cb22-145" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;summary&gt;&lt;i&gt;</span>Click here to see how to do this in base R<span class="kw">&lt;/i&gt;&lt;/summary&gt;</span></span>
<span id="cb22-146"><a href="#cb22-146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-147"><a href="#cb22-147" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;br&gt;</span></span>
<span id="cb22-148"><a href="#cb22-148" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-149"><a href="#cb22-149" aria-hidden="true" tabindex="-1"></a><span class="in">```{=html}</span></span>
<span id="cb22-150"><a href="#cb22-150" aria-hidden="true" tabindex="-1"></a><span class="in">&lt;pre&gt;&lt;code&gt;# Read in the functional analysis results&lt;br&gt;</span></span>
<span id="cb22-151"><a href="#cb22-151" aria-hidden="true" tabindex="-1"></a><span class="in">functional_GO_results &lt;- read.delim(file = "data/gprofiler_results_Mov10oe.tsv", sep = "\t" )&lt;/code&gt;&lt;br&gt;</span></span>
<span id="cb22-152"><a href="#cb22-152" aria-hidden="true" tabindex="-1"></a><span class="in">&lt;code&gt;# Take a look at the results</span></span>
<span id="cb22-153"><a href="#cb22-153" aria-hidden="true" tabindex="-1"></a><span class="in">functional_GO_results&lt;/code&gt;&lt;/pre&gt;</span></span>
<span id="cb22-154"><a href="#cb22-154" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-155"><a href="#cb22-155" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;br&gt;</span></span>
<span id="cb22-156"><a href="#cb22-156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-157"><a href="#cb22-157" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/details&gt;</span></span>
<span id="cb22-158"><a href="#cb22-158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-159"><a href="#cb22-159" aria-hidden="true" tabindex="-1"></a>Notice that the results were automatically read in as a tibble and the output gives the number of rows, columns and the data type for each of the columns.</span>
<span id="cb22-160"><a href="#cb22-160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-161"><a href="#cb22-161" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb22-162"><a href="#cb22-162" aria-hidden="true" tabindex="-1"></a>A large number of tidyverse functions will work with both tibbles and dataframes, and the data structure of the output will be identical to the input. However, there are some functions that will return a tibble (without row names), whether or not a tibble or dataframe is provided.</span>
<span id="cb22-163"><a href="#cb22-163" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb22-164"><a href="#cb22-164" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-165"><a href="#cb22-165" aria-hidden="true" tabindex="-1"></a><span class="fu">## 2. Extract only the GO biological processes (BP) of interest</span></span>
<span id="cb22-166"><a href="#cb22-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-167"><a href="#cb22-167" aria-hidden="true" tabindex="-1"></a>Now that we have our data, we will need to wrangle it into a format ready for plotting. For all of our data wrangling steps we will be using tools from the <span class="co">[</span><span class="ot">dplyr</span><span class="co">](http://dplyr.tidyverse.org/)</span> package, which is a swiss-army knife for data wrangling of data frames.</span>
<span id="cb22-168"><a href="#cb22-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-169"><a href="#cb22-169" aria-hidden="true" tabindex="-1"></a>To extract the biological processes of interest, we only want those rows where the <span class="in">`domain`</span> is equal to <span class="in">`BP`</span>, which we can do using the <span class="in">`filter()`</span> function.</span>
<span id="cb22-170"><a href="#cb22-170" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-171"><a href="#cb22-171" aria-hidden="true" tabindex="-1"></a>To filter rows of a data frame/tibble based on values in different columns, we give a logical expression as input to the <span class="in">`filter()`</span> function to return those rows for which the expression is TRUE.</span>
<span id="cb22-172"><a href="#cb22-172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-173"><a href="#cb22-173" aria-hidden="true" tabindex="-1"></a>Now let's return only those rows that have a <span class="in">`domain`</span> of <span class="in">`BP`</span>:</span>
<span id="cb22-174"><a href="#cb22-174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-177"><a href="#cb22-177" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb22-178"><a href="#cb22-178" aria-hidden="true" tabindex="-1"></a><span class="co"># Return only GO biological processes</span></span>
<span id="cb22-179"><a href="#cb22-179" aria-hidden="true" tabindex="-1"></a>bp_oe <span class="ot">&lt;-</span> functional_GO_results <span class="sc">%&gt;%</span></span>
<span id="cb22-180"><a href="#cb22-180" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(domain <span class="sc">==</span> <span class="st">"BP"</span>)</span>
<span id="cb22-181"><a href="#cb22-181" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb22-182"><a href="#cb22-182" aria-hidden="true" tabindex="-1"></a>bp_oe</span>
<span id="cb22-183"><a href="#cb22-183" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-184"><a href="#cb22-184" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-185"><a href="#cb22-185" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;details&gt;</span></span>
<span id="cb22-186"><a href="#cb22-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-187"><a href="#cb22-187" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;summary&gt;&lt;i&gt;</span>Click here to see how to do this in base R<span class="kw">&lt;/i&gt;&lt;/summary&gt;</span></span>
<span id="cb22-188"><a href="#cb22-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-189"><a href="#cb22-189" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;br&gt;</span></span>
<span id="cb22-190"><a href="#cb22-190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-191"><a href="#cb22-191" aria-hidden="true" tabindex="-1"></a><span class="in">```{=html}</span></span>
<span id="cb22-192"><a href="#cb22-192" aria-hidden="true" tabindex="-1"></a><span class="in">&lt;pre&gt;&lt;code&gt;# Return only GO biological processes</span></span>
<span id="cb22-193"><a href="#cb22-193" aria-hidden="true" tabindex="-1"></a><span class="in">idx &lt;- functional_GO_results$domain == "BP"</span></span>
<span id="cb22-194"><a href="#cb22-194" aria-hidden="true" tabindex="-1"></a><span class="in">bp_oe2 &lt;- functional_GO_results[idx,]&lt;/code&gt;&lt;br&gt;</span></span>
<span id="cb22-195"><a href="#cb22-195" aria-hidden="true" tabindex="-1"></a><span class="in">&lt;code&gt;View(bp_oe)&lt;/code&gt;&lt;/pre&gt;</span></span>
<span id="cb22-196"><a href="#cb22-196" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-197"><a href="#cb22-197" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;br&gt;</span></span>
<span id="cb22-198"><a href="#cb22-198" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-199"><a href="#cb22-199" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/details&gt;</span></span>
<span id="cb22-200"><a href="#cb22-200" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-201"><a href="#cb22-201" aria-hidden="true" tabindex="-1"></a>Now we have returned only those rows with a <span class="in">`domain`</span> of <span class="in">`BP`</span>. **How have the dimensions of our results changed?**</span>
<span id="cb22-202"><a href="#cb22-202" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-203"><a href="#cb22-203" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb22-204"><a href="#cb22-204" aria-hidden="true" tabindex="-1"></a><span class="fu"># Exercises</span></span>
<span id="cb22-205"><a href="#cb22-205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-206"><a href="#cb22-206" aria-hidden="true" tabindex="-1"></a>We would like to perform an additional round of filtering to only keep the most specific GO terms.</span>
<span id="cb22-207"><a href="#cb22-207" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-208"><a href="#cb22-208" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>For <span class="in">`bp_oe`</span>, use the <span class="in">`filter()`</span> function to only keep those rows where the <span class="in">`relative.depth`</span> is greater than 4.</span>
<span id="cb22-209"><a href="#cb22-209" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Save output to overwrite our <span class="in">`bp_oe`</span> variable.</span>
<span id="cb22-210"><a href="#cb22-210" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb22-211"><a href="#cb22-211" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-212"><a href="#cb22-212" aria-hidden="true" tabindex="-1"></a><span class="fu">## 3. Select only the columns needed for visualization</span></span>
<span id="cb22-213"><a href="#cb22-213" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-214"><a href="#cb22-214" aria-hidden="true" tabindex="-1"></a>For visualization purposes, we are only interested in the columns related to the GO terms, the significance of the terms, and information about the number of genes associated with the terms.</span>
<span id="cb22-215"><a href="#cb22-215" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-216"><a href="#cb22-216" aria-hidden="true" tabindex="-1"></a>To extract columns from a data frame/tibble we can use the <span class="in">`select()`</span> function. In contrast to base R, we do not need to put the column names in quotes for selection.</span>
<span id="cb22-217"><a href="#cb22-217" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-220"><a href="#cb22-220" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb22-221"><a href="#cb22-221" aria-hidden="true" tabindex="-1"></a><span class="co"># Selecting columns to keep</span></span>
<span id="cb22-222"><a href="#cb22-222" aria-hidden="true" tabindex="-1"></a>bp_oe <span class="ot">&lt;-</span> bp_oe <span class="sc">%&gt;%</span></span>
<span id="cb22-223"><a href="#cb22-223" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(term.id, term.name, p.value, query.size, term.size, overlap.size, intersection)</span>
<span id="cb22-224"><a href="#cb22-224" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-225"><a href="#cb22-225" aria-hidden="true" tabindex="-1"></a>bp_oe</span>
<span id="cb22-226"><a href="#cb22-226" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-227"><a href="#cb22-227" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-228"><a href="#cb22-228" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;details&gt;</span></span>
<span id="cb22-229"><a href="#cb22-229" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-230"><a href="#cb22-230" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;summary&gt;&lt;i&gt;</span>Click here to see how to do this in base R<span class="kw">&lt;/i&gt;&lt;/summary&gt;</span></span>
<span id="cb22-231"><a href="#cb22-231" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-232"><a href="#cb22-232" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;br&gt;</span></span>
<span id="cb22-233"><a href="#cb22-233" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-234"><a href="#cb22-234" aria-hidden="true" tabindex="-1"></a><span class="in">```{=html}</span></span>
<span id="cb22-235"><a href="#cb22-235" aria-hidden="true" tabindex="-1"></a><span class="in">&lt;pre&gt;&lt;code&gt;# Selecting columns to keep</span></span>
<span id="cb22-236"><a href="#cb22-236" aria-hidden="true" tabindex="-1"></a><span class="in">bp_oe &lt;- bp_oe[, c("term.id", "term.name", "p.value", "query.size", "term.size", "overlap.size", "intersection")]&lt;/code&gt;&lt;br&gt;</span></span>
<span id="cb22-237"><a href="#cb22-237" aria-hidden="true" tabindex="-1"></a><span class="in">&lt;code&gt;View(bp_oe)&lt;/code&gt;&lt;/pre&gt;</span></span>
<span id="cb22-238"><a href="#cb22-238" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-239"><a href="#cb22-239" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;br&gt;</span></span>
<span id="cb22-240"><a href="#cb22-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-241"><a href="#cb22-241" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/details&gt;</span></span>
<span id="cb22-242"><a href="#cb22-242" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-243"><a href="#cb22-243" aria-hidden="true" tabindex="-1"></a>The <span class="in">`select()`</span> function also allows for negative selection. So we could have alternately removed columns with negative selection. Note that we need to put the column names inside of the combine (<span class="in">`c()`</span>) function with a <span class="in">`-`</span> preceding it for this functionality.</span>
<span id="cb22-244"><a href="#cb22-244" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-247"><a href="#cb22-247" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb22-248"><a href="#cb22-248" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb22-249"><a href="#cb22-249" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb22-250"><a href="#cb22-250" aria-hidden="true" tabindex="-1"></a><span class="co"># Selecting columns to remove</span></span>
<span id="cb22-251"><a href="#cb22-251" aria-hidden="true" tabindex="-1"></a>bp_oe <span class="ot">&lt;-</span> bp_oe <span class="sc">%&gt;%</span></span>
<span id="cb22-252"><a href="#cb22-252" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(query.number, significant, recall, precision, </span>
<span id="cb22-253"><a href="#cb22-253" aria-hidden="true" tabindex="-1"></a>              subgraph.number, relative.depth, domain))</span>
<span id="cb22-254"><a href="#cb22-254" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-255"><a href="#cb22-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-256"><a href="#cb22-256" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;details&gt;</span></span>
<span id="cb22-257"><a href="#cb22-257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-258"><a href="#cb22-258" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;summary&gt;&lt;i&gt;</span>Click here to see how to do this in base R<span class="kw">&lt;/i&gt;&lt;/summary&gt;</span></span>
<span id="cb22-259"><a href="#cb22-259" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-260"><a href="#cb22-260" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;br&gt;</span></span>
<span id="cb22-261"><a href="#cb22-261" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-262"><a href="#cb22-262" aria-hidden="true" tabindex="-1"></a><span class="in">```{=html}</span></span>
<span id="cb22-263"><a href="#cb22-263" aria-hidden="true" tabindex="-1"></a><span class="in">&lt;pre&gt;&lt;code&gt;# DO NOT RUN&lt;/code&gt;</span></span>
<span id="cb22-264"><a href="#cb22-264" aria-hidden="true" tabindex="-1"></a><span class="in">&lt;code&gt;# Selecting columns to remove</span></span>
<span id="cb22-265"><a href="#cb22-265" aria-hidden="true" tabindex="-1"></a><span class="in">idx &lt;- !(colnames(bp_oe) %in% c("query.number", "significant", "recall", "precision", "subgraph.number", "relative.depth", "domain"))</span></span>
<span id="cb22-266"><a href="#cb22-266" aria-hidden="true" tabindex="-1"></a><span class="in">bp_oe &lt;- bp_oe[, idx]&lt;/code&gt;&lt;/pre&gt;</span></span>
<span id="cb22-267"><a href="#cb22-267" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-268"><a href="#cb22-268" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;br&gt;</span></span>
<span id="cb22-269"><a href="#cb22-269" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-270"><a href="#cb22-270" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/details&gt;</span></span>
<span id="cb22-271"><a href="#cb22-271" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-272"><a href="#cb22-272" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;img</span> <span class="er">src</span><span class="ot">=</span><span class="st">"../img/bp_oe_selection.png"</span> <span class="er">width</span><span class="ot">=</span><span class="st">"800"</span><span class="kw">/&gt;</span></span>
<span id="cb22-273"><a href="#cb22-273" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-274"><a href="#cb22-274" aria-hidden="true" tabindex="-1"></a><span class="fu">## 4. Order GO processes by significance (adjusted p-values)</span></span>
<span id="cb22-275"><a href="#cb22-275" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-276"><a href="#cb22-276" aria-hidden="true" tabindex="-1"></a>Now that we have only the rows and columns of interest, let's arrange these by significance, which is denoted by the adjusted p-value.</span>
<span id="cb22-277"><a href="#cb22-277" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-278"><a href="#cb22-278" aria-hidden="true" tabindex="-1"></a>Let's sort the rows by adjusted p-value with the <span class="in">`arrange()`</span> function.</span>
<span id="cb22-279"><a href="#cb22-279" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-282"><a href="#cb22-282" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb22-283"><a href="#cb22-283" aria-hidden="true" tabindex="-1"></a><span class="co"># Order by adjusted p-value ascending</span></span>
<span id="cb22-284"><a href="#cb22-284" aria-hidden="true" tabindex="-1"></a>bp_oe <span class="ot">&lt;-</span> bp_oe <span class="sc">%&gt;%</span></span>
<span id="cb22-285"><a href="#cb22-285" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(p.value)</span>
<span id="cb22-286"><a href="#cb22-286" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-287"><a href="#cb22-287" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-288"><a href="#cb22-288" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;details&gt;</span></span>
<span id="cb22-289"><a href="#cb22-289" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-290"><a href="#cb22-290" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;summary&gt;&lt;i&gt;</span>Click here to see how to do this in base R<span class="kw">&lt;/i&gt;&lt;/summary&gt;</span></span>
<span id="cb22-291"><a href="#cb22-291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-292"><a href="#cb22-292" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;br&gt;</span></span>
<span id="cb22-293"><a href="#cb22-293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-294"><a href="#cb22-294" aria-hidden="true" tabindex="-1"></a><span class="in">```{=html}</span></span>
<span id="cb22-295"><a href="#cb22-295" aria-hidden="true" tabindex="-1"></a><span class="in">&lt;pre&gt;&lt;code&gt;# Order by adjusted p-value ascending</span></span>
<span id="cb22-296"><a href="#cb22-296" aria-hidden="true" tabindex="-1"></a><span class="in">idx &lt;- order(bp_oe$p.value)</span></span>
<span id="cb22-297"><a href="#cb22-297" aria-hidden="true" tabindex="-1"></a><span class="in">bp_oe &lt;- bp_oe[idx,]&lt;/code&gt;&lt;/pre&gt;</span></span>
<span id="cb22-298"><a href="#cb22-298" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-299"><a href="#cb22-299" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;br&gt;</span></span>
<span id="cb22-300"><a href="#cb22-300" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-301"><a href="#cb22-301" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/details&gt;</span></span>
<span id="cb22-302"><a href="#cb22-302" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-303"><a href="#cb22-303" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb22-304"><a href="#cb22-304" aria-hidden="true" tabindex="-1"></a>If you wanted to arrange in descending order, then you could have run the following instead:</span>
<span id="cb22-305"><a href="#cb22-305" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-308"><a href="#cb22-308" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb22-309"><a href="#cb22-309" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb22-310"><a href="#cb22-310" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb22-311"><a href="#cb22-311" aria-hidden="true" tabindex="-1"></a><span class="co"># Order by adjusted p-value descending</span></span>
<span id="cb22-312"><a href="#cb22-312" aria-hidden="true" tabindex="-1"></a>bp_oe <span class="ot">&lt;-</span> bp_oe <span class="sc">%&gt;%</span></span>
<span id="cb22-313"><a href="#cb22-313" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(p.value))</span>
<span id="cb22-314"><a href="#cb22-314" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-315"><a href="#cb22-315" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-316"><a href="#cb22-316" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;details&gt;</span></span>
<span id="cb22-317"><a href="#cb22-317" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-318"><a href="#cb22-318" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;summary&gt;&lt;i&gt;</span>Click here to see how to do this in base R<span class="kw">&lt;/i&gt;&lt;/summary&gt;</span></span>
<span id="cb22-319"><a href="#cb22-319" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-320"><a href="#cb22-320" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;br&gt;</span></span>
<span id="cb22-321"><a href="#cb22-321" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-322"><a href="#cb22-322" aria-hidden="true" tabindex="-1"></a><span class="in">```{=html}</span></span>
<span id="cb22-323"><a href="#cb22-323" aria-hidden="true" tabindex="-1"></a><span class="in">&lt;pre&gt;&lt;code&gt;# DO NOT RUN&lt;/code&gt;</span></span>
<span id="cb22-324"><a href="#cb22-324" aria-hidden="true" tabindex="-1"></a><span class="in">&lt;code&gt;# Order by adjusted p-value descending</span></span>
<span id="cb22-325"><a href="#cb22-325" aria-hidden="true" tabindex="-1"></a><span class="in">idx &lt;- order(bp_oe$p.value, decreasing = TRUE)</span></span>
<span id="cb22-326"><a href="#cb22-326" aria-hidden="true" tabindex="-1"></a><span class="in">bp_oe &lt;- bp_oe[idx,]&lt;/code&gt;&lt;/pre&gt;</span></span>
<span id="cb22-327"><a href="#cb22-327" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-328"><a href="#cb22-328" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/details&gt;</span></span>
<span id="cb22-329"><a href="#cb22-329" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb22-330"><a href="#cb22-330" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-331"><a href="#cb22-331" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb22-332"><a href="#cb22-332" aria-hidden="true" tabindex="-1"></a>Ordering variables in <span class="in">`ggplot2`</span> is a bit different. <span class="co">[</span><span class="ot">This post</span><span class="co">](https://www.r-graph-gallery.com/267-reorder-a-variable-in-ggplot2.html)</span> introduces a few ways of ordering variables in a plot.</span>
<span id="cb22-333"><a href="#cb22-333" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb22-334"><a href="#cb22-334" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-335"><a href="#cb22-335" aria-hidden="true" tabindex="-1"></a><span class="fu">## 5. Rename columns to be more intuitive</span></span>
<span id="cb22-336"><a href="#cb22-336" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-337"><a href="#cb22-337" aria-hidden="true" tabindex="-1"></a>While not necessary for our visualization, renaming columns more intuitively can help with our understanding of the data using the <span class="in">`rename()`</span> function. The syntax is <span class="in">`new_name`</span> = <span class="in">`old_name`</span>.</span>
<span id="cb22-338"><a href="#cb22-338" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-339"><a href="#cb22-339" aria-hidden="true" tabindex="-1"></a>Let's rename the <span class="in">`term.id`</span> and <span class="in">`term.name`</span> columns.</span>
<span id="cb22-340"><a href="#cb22-340" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-343"><a href="#cb22-343" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb22-344"><a href="#cb22-344" aria-hidden="true" tabindex="-1"></a><span class="co"># Provide better names for columns</span></span>
<span id="cb22-345"><a href="#cb22-345" aria-hidden="true" tabindex="-1"></a>bp_oe <span class="ot">&lt;-</span> bp_oe <span class="sc">%&gt;%</span> </span>
<span id="cb22-346"><a href="#cb22-346" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">GO_id =</span> term.id, </span>
<span id="cb22-347"><a href="#cb22-347" aria-hidden="true" tabindex="-1"></a>                <span class="at">GO_term =</span> term.name)</span>
<span id="cb22-348"><a href="#cb22-348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-349"><a href="#cb22-349" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(bp_oe)</span>
<span id="cb22-350"><a href="#cb22-350" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-351"><a href="#cb22-351" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-352"><a href="#cb22-352" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;details&gt;</span></span>
<span id="cb22-353"><a href="#cb22-353" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-354"><a href="#cb22-354" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;summary&gt;&lt;i&gt;</span>Click here to see how to do this in base R<span class="kw">&lt;/i&gt;&lt;/summary&gt;</span></span>
<span id="cb22-355"><a href="#cb22-355" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-356"><a href="#cb22-356" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;br&gt;</span></span>
<span id="cb22-357"><a href="#cb22-357" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-358"><a href="#cb22-358" aria-hidden="true" tabindex="-1"></a><span class="in">```{=html}</span></span>
<span id="cb22-359"><a href="#cb22-359" aria-hidden="true" tabindex="-1"></a><span class="in">&lt;pre&gt;&lt;code&gt;# Provide better names for columns</span></span>
<span id="cb22-360"><a href="#cb22-360" aria-hidden="true" tabindex="-1"></a><span class="in">colnames(bp_oe)[colnames(bp_oe) == "term.id"] &lt;- "GO_id"</span></span>
<span id="cb22-361"><a href="#cb22-361" aria-hidden="true" tabindex="-1"></a><span class="in">colnames(bp_oe)[colnames(bp_oe) == "term.name"] &lt;- "GO_term"&lt;/code&gt;&lt;/pre&gt;</span></span>
<span id="cb22-362"><a href="#cb22-362" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-363"><a href="#cb22-363" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;br&gt;</span></span>
<span id="cb22-364"><a href="#cb22-364" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-365"><a href="#cb22-365" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/details&gt;</span></span>
<span id="cb22-366"><a href="#cb22-366" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-367"><a href="#cb22-367" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb22-368"><a href="#cb22-368" aria-hidden="true" tabindex="-1"></a>In the case of two packages with identical function names, you can use <span class="in">`::`</span> with the package name before and the function name after (e.g <span class="in">`stats::filter()`</span>) to ensure that the correct function is implemented. The <span class="in">`::`</span> can also be used to bring in a function from a library without loading it first.</span>
<span id="cb22-369"><a href="#cb22-369" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-370"><a href="#cb22-370" aria-hidden="true" tabindex="-1"></a>In the example above, we wanted to use the <span class="in">`rename()`</span> function specifically from the <span class="in">`dplyr`</span> package, and not any of the other packages (or base R) which may have the <span class="in">`rename()`</span> function.</span>
<span id="cb22-371"><a href="#cb22-371" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb22-372"><a href="#cb22-372" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-373"><a href="#cb22-373" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb22-374"><a href="#cb22-374" aria-hidden="true" tabindex="-1"></a><span class="fu"># Exercises</span></span>
<span id="cb22-375"><a href="#cb22-375" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-376"><a href="#cb22-376" aria-hidden="true" tabindex="-1"></a>Rename the <span class="in">`intersection`</span> column to <span class="in">`genes`</span> to reflect the fact that these are the DE genes associated with the GO process.</span>
<span id="cb22-377"><a href="#cb22-377" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb22-378"><a href="#cb22-378" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-379"><a href="#cb22-379" aria-hidden="true" tabindex="-1"></a><span class="fu">## 6. Create additional metrics for plotting (e.g. gene ratios)</span></span>
<span id="cb22-380"><a href="#cb22-380" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-381"><a href="#cb22-381" aria-hidden="true" tabindex="-1"></a>Finally, before we plot our data, we need to create a couple of additional metrics. The <span class="in">`mutate()`</span> function enables you to create a new column from an existing column.</span>
<span id="cb22-382"><a href="#cb22-382" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-383"><a href="#cb22-383" aria-hidden="true" tabindex="-1"></a>Let's generate gene ratios to reflect the number of DE genes associated with each GO process relative to the total number of DE genes.</span>
<span id="cb22-384"><a href="#cb22-384" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-387"><a href="#cb22-387" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb22-388"><a href="#cb22-388" aria-hidden="true" tabindex="-1"></a><span class="co"># Create gene ratio column based on other columns in dataset</span></span>
<span id="cb22-389"><a href="#cb22-389" aria-hidden="true" tabindex="-1"></a>bp_oe <span class="ot">&lt;-</span> bp_oe <span class="sc">%&gt;%</span></span>
<span id="cb22-390"><a href="#cb22-390" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">gene_ratio =</span> overlap.size <span class="sc">/</span> query.size)</span>
<span id="cb22-391"><a href="#cb22-391" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(bp_oe)</span>
<span id="cb22-392"><a href="#cb22-392" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-393"><a href="#cb22-393" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-394"><a href="#cb22-394" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;details&gt;</span></span>
<span id="cb22-395"><a href="#cb22-395" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-396"><a href="#cb22-396" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;summary&gt;&lt;i&gt;</span>Click here to see how to do this in base R<span class="kw">&lt;/i&gt;&lt;/summary&gt;</span></span>
<span id="cb22-397"><a href="#cb22-397" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-398"><a href="#cb22-398" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;br&gt;</span></span>
<span id="cb22-399"><a href="#cb22-399" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-400"><a href="#cb22-400" aria-hidden="true" tabindex="-1"></a><span class="in">```{=html}</span></span>
<span id="cb22-401"><a href="#cb22-401" aria-hidden="true" tabindex="-1"></a><span class="in">&lt;pre&gt;&lt;code&gt;# Create gene ratio column based on other columns in dataset</span></span>
<span id="cb22-402"><a href="#cb22-402" aria-hidden="true" tabindex="-1"></a><span class="in">bp_oe &lt;- cbind(bp_oe, gene_ratio = bp_oe$overlap.size / bp_oe$query.size)&lt;/code&gt;&lt;/pre&gt;</span></span>
<span id="cb22-403"><a href="#cb22-403" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-404"><a href="#cb22-404" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;br&gt;</span></span>
<span id="cb22-405"><a href="#cb22-405" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-406"><a href="#cb22-406" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/details&gt;</span></span>
<span id="cb22-407"><a href="#cb22-407" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-408"><a href="#cb22-408" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb22-409"><a href="#cb22-409" aria-hidden="true" tabindex="-1"></a><span class="fu"># Exercises</span></span>
<span id="cb22-410"><a href="#cb22-410" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-411"><a href="#cb22-411" aria-hidden="true" tabindex="-1"></a>Create a column in <span class="in">`bp_oe`</span> called <span class="in">`term_percent`</span> to determine the percent of DE genes associated with the GO term relative to the total number of genes associated with the GO term (<span class="in">`overlap.size`</span> / <span class="in">`term.size`</span>)</span>
<span id="cb22-412"><a href="#cb22-412" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb22-413"><a href="#cb22-413" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-414"><a href="#cb22-414" aria-hidden="true" tabindex="-1"></a>Our final data for plotting should look like the table below:</span>
<span id="cb22-415"><a href="#cb22-415" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-416"><a href="#cb22-416" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;img</span> <span class="er">src</span><span class="ot">=</span><span class="st">"../img/bp_oe_ready_to_plot.png"</span> <span class="er">width</span><span class="ot">=</span><span class="st">"900"</span><span class="kw">/&gt;</span></span>
<span id="cb22-417"><a href="#cb22-417" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-418"><a href="#cb22-418" aria-hidden="true" tabindex="-1"></a><span class="fu">## Next steps</span></span>
<span id="cb22-419"><a href="#cb22-419" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-420"><a href="#cb22-420" aria-hidden="true" tabindex="-1"></a>Now that we have our results ready for plotting, we can use the <span class="co">[</span><span class="ot">ggplot2</span><span class="co">](https://ggplot2.tidyverse.org)</span> package to plot our results. If you are interested, you can follow <span class="co">[</span><span class="ot">this lesson</span><span class="co">](https://hbctraining.github.io/Training-modules/Tidyverse_ggplot2/lessons/03_ggplot2.html)</span> and dive into how to use <span class="in">`ggplot2`</span> to create the plots with this dataset.</span>
<span id="cb22-421"><a href="#cb22-421" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-422"><a href="#cb22-422" aria-hidden="true" tabindex="-1"></a><span class="fu">### Additional resources</span></span>
<span id="cb22-423"><a href="#cb22-423" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-424"><a href="#cb22-424" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="co">[</span><span class="ot">R for Data Science</span><span class="co">](http://r4ds.had.co.nz)</span></span>
<span id="cb22-425"><a href="#cb22-425" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="co">[</span><span class="ot">teach the tidyverse</span><span class="co">](http://varianceexplained.org/r/teach-tidyverse/)</span></span>
<span id="cb22-426"><a href="#cb22-426" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="co">[</span><span class="ot">tidy style guide</span><span class="co">](http://style.tidyverse.org/)</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">This lesson has been developed by members of the teaching team at the <a href="http://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core (HBC)</a>. <br> These are open access materials distributed under the terms of the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license</a> (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited. <br> Some materials used are adapted from work that is Copyright © <a href="http://datacarpentry.org/">Data Carpentry</a>. All Data Carpentry instructional material is made available under the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license</a> (CC BY 4.0).</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>



</body></html>