<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Mary Piper, Meeta Mistry, Radhika Khetani">

<title>Plotting and data visualization in R – Introduction to R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-e5cb4ee4193d6a2f28a249434e39fb42.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../css/styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Introduction to R</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../lessons/schedule.html"> 
<span class="menu-text">Schedule</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://bioinformatics.sph.harvard.edu/"> 
<span class="menu-text">HBC</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/hbctraining/intro_r_rmd"> 
<span class="menu-text">GitHub</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="mailto:hbctraining@hsph.harvard.edu"> 
<span class="menu-text">Contact us</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link active" data-scroll-target="#learning-objectives">Learning Objectives</a></li>
  <li><a href="#data-visualization-with-ggplot2" id="toc-data-visualization-with-ggplot2" class="nav-link" data-scroll-target="#data-visualization-with-ggplot2">Data Visualization with <code>ggplot2</code></a>
  <ul class="collapse">
  <li><a href="#customizing-data-point-colors" id="toc-customizing-data-point-colors" class="nav-link" data-scroll-target="#customizing-data-point-colors">Customizing data point colors</a></li>
  <li><a href="#exporting-figures-to-file" id="toc-exporting-figures-to-file" class="nav-link" data-scroll-target="#exporting-figures-to-file">Exporting figures to file</a></li>
  <li><a href="#consistent-formatting-using-custom-functions" id="toc-consistent-formatting-using-custom-functions" class="nav-link" data-scroll-target="#consistent-formatting-using-custom-functions">Consistent formatting using custom functions</a></li>
  </ul></li>
  <li><a href="#resources" id="toc-resources" class="nav-link" data-scroll-target="#resources">Resources</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Plotting and data visualization in R</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Mary Piper, Meeta Mistry, Radhika Khetani </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">Invalid Date</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<p>Approximate time: 90 minutes</p>
<section id="learning-objectives" class="level2">
<h2 class="anchored" data-anchor-id="learning-objectives">Learning Objectives</h2>
<ul>
<li>Explain the grammar of graphics syntax used by ggplot2</li>
<li>Determine how to plot different types of graphs with ggplot2 depending on the number and type of variables</li>
<li>Export plots for use outside of the R environment.</li>
</ul>
</section>
<section id="data-visualization-with-ggplot2" class="level2">
<h2 class="anchored" data-anchor-id="data-visualization-with-ggplot2">Data Visualization with <code>ggplot2</code></h2>
<p>When we are working with large sets of numbers it can be useful to display that information graphically to gain more insight. Visualization deserves an entire course of its own (there is that much to know!). In this lesson we will be plotting with the popular package <a href="http://docs.ggplot2.org/"><code>ggplot2</code></a>.</p>
<p>More recently, R users have moved away from base graphic options towards <code>ggplot2</code> since it offers a lot more functionality as compared to the base R plotting functions. The <code>ggplot2</code> syntax takes some getting used to, but once you get it, you will find it’s extremely powerful and flexible. We will start with drawing a simple X-Y scatterplot of <code>gene_ratio</code> versus <code>GO_term</code> from the <code>bp_oe</code> tibble. <code>ggplot2</code> expects that the information being plotted is contained in a data frame or tibble (data frame-like).</p>
<p>We would typically start by loading the <code>ggplot2</code> library, but it is a part of the <code>tidyverse</code> suite, so it was loaded in the last lesson.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"data/GOs_for_plotting.RData"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>ggplot2 syntax:</strong> To <strong>initialize the basic graph structure</strong> with this package we have to use the <code>ggplot()</code> function, then we add “layers” to it using the <code>+</code> operator. The idea is that you create a basic plot first, then additional functions are added on to build the final plot.</p>
<p>Let’s start by first subsetting the <code>bp_oe</code> dataframe to only contain the top 30 most significant GO terms:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load ggplot2 package</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset data frame</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>bp_plot <span class="ot">&lt;-</span> bp_oe[<span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now let’s try making a plot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># What happens if we don't provide it a geom object?</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You get a blank plot, because you need to <strong>specify layers</strong> using the <code>+</code> operator.</p>
<p>One type of layer is <strong>geometric objects</strong>. This is a mandatory layer and it specifies what type of plot you are interested in making. Examples include:</p>
<ul>
<li>points (<code>geom_point()</code>, <code>geom_jitter()</code> for scatter plots, dot plots, etc)</li>
<li>lines (<code>geom_line()</code>, for time series, trend lines, etc)</li>
<li>boxplot (<code>geom_boxplot()</code>, for, well, boxplots!)</li>
</ul>
<p>For a more exhaustive list on all possible geometric objects and when to use them check out <a href="http://rpubs.com/hadley/ggplot2-layers">Hadley Wickham’s RPubs</a> or the <a href="https://rstudio.github.io/cheatsheets/html/data-visualization.html">RStudio cheatsheet</a>.</p>
<p>A plot <strong>must have at least one <code>geom</code></strong>, and there can be multiple <em>complementary</em> <code>geom</code>s; there is no upper limit.</p>
<p>Let’s add a <code>geom</code> to make a scatter plot, i.e.&nbsp;the <code>geom_point()</code> function</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Provide the plot a dataset and a type of plot</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You will find that even though we have added a layer by specifying <code>geom_point</code>, we get an error. This is because each type of geom usually has a required set of <strong>aesthetics</strong>. Aesthetic mappings are set with the <code>aes()</code> function which can be nested within the <code>geom</code> function and can be set inside <code>geom_point()</code> to be specifically applied to that layer. If we supplied aesthetics within <code>ggplot()</code>, they will be used as defaults for every layer. Below are some examples of what is categorized as aesthetics in the ggplot2 context:</p>
<ul>
<li>position (i.e.&nbsp;columns to be used for the x and y axes)</li>
<li>color (“outside” color of the data point/bar being plotted)</li>
<li>fill (“inside” color of the data point/bar being plotted)</li>
<li>shape (of data points)</li>
<li>linetype</li>
<li>size (of data points)</li>
</ul>
<p>To start, we will specify the columns for the x- and y-axis since <code>geom_point()</code> requires the most basic information about a scatterplot, i.e.&nbsp;what you want to plot on the x and y axes.</p>
<p>Typically, a scatterplot is used to illustrate the relationship between two numeric variables. The x-axis represents the independent variable and the y-axis represents the dependent variable. We can test this out using two numeric columns from our <code>bp_plot</code> tibble:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Provide the plot a dataset, a type of plot and an aesthetics function</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> overlap.size, <span class="at">y =</span> p.value))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Aside_ggplot2_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>However, instead of a scatterplot with numeric values on both axes, we would like to create a dotplot for visualizing the top 30 functional categories in our dataset, and how prevalent they are. Basically, we want a dotplot for visualizing functional analysis data, which plots the gene ratio values on the x-axis and the GO terms on the y-axis.</p>
<p>Let’s see what happens when we add a non-numeric value to the y-axis and change the x-axis to the “gene_ratio” column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_ratio, <span class="at">y =</span> GO_term))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Aside_ggplot2_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Now that we have the required aesthetics, let’s add some extras like color to the plot. Let’s say we wanted <em>to quickly visualize significance of the GO terms</em> in the plot, we can <strong><code>color</code> the points on the plot based on p-values</strong>, by specifying the column header.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_ratio, <span class="at">y =</span> GO_term, <span class="at">color =</span> p.value))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Aside_ggplot2_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>You will notice that there are a default set of colors that will be used so we do not have to specify which colors to use. Also, the <strong>legend has been conveniently plotted for us!</strong></p>
<p>Alternatively, we could color number of DE genes associated with each term (<code>overlap.size</code>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_ratio, <span class="at">y =</span> GO_term, <span class="at">color =</span> overlap.size))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Aside_ggplot2_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Moving forward, we are going to stick with coloring the dots based on the p.value column. Let’s explore some of the other arguments that can be specified in the <code>geom</code> layer.</p>
<p>To modify the <strong>size of the data points</strong> we can use the <code>size</code> argument. * If we add <code>size</code> inside <code>aes()</code> we could assign a numeric column to it and the size of the data points would change according to that column. * However, if we add <code>size</code> inside the <code>geom_point()</code> but outside <code>aes()</code> we can’t assign a column to it, instead we have to give it a numeric value. This use of <code>size</code> will uniformly change the size of all the data points.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>This is true for several arguments, including <code>color</code>, <code>shape</code> etc. E.g. we can change all shapes to square by adding this argument to be outside the <code>aes()</code> function; if we put the argument inside the <code>aes()</code> function we could change the shape according to a (categorical) variable in our data frame or tibble.</p>
</div>
</div>
<p>We have decided that we want to change the size of all the data point to a uniform size instead of typing it to a numeric column in the input tibble. Add in the <code>size</code> argument by specifying a number for the size of the data point:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_ratio, <span class="at">y =</span> GO_term, , <span class="at">color =</span> p.value), </span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">size =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Aside_ggplot2_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The size of the points is personal preference, and you may need to play around with the parameter to decide which size is best. That seems a bit too small, so we can try out a slightly larger size.</p>
</div>
</div>
<p>As we do that, let’s see how we can change the shape of the data point. Different shapes are available, as detailed in the <a href="https://rstudio.github.io/cheatsheets/html/data-visualization.html">RStudio ggplot2 cheatsheet</a>. Let’s explore this parameter by changing all of the points to squares:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_ratio, <span class="at">y =</span> GO_term, , <span class="at">color =</span> p.value), </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">size =</span> <span class="dv">2</span>, </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">shape =</span> <span class="st">"square"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Aside_ggplot2_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Now we can start updating the plot to suit our preferences for how we want the data displayed. The labels on the x- and y-axis are also quite small and not very descriptive. To change their size and labeling, we need to add additional <strong>theme layers</strong>. The ggplot2 <code>theme()</code> system handles modification of non-data plot elements such as:</p>
<ul>
<li>Axis label aesthetics</li>
<li>Plot background</li>
<li>Facet label backround</li>
<li>Legend appearance</li>
</ul>
<p>There are built-in themes that we can use (i.e.&nbsp;<code>theme_bw()</code>) that mostly change the background/foreground colors, by adding it as additional layer. Alternatively, we can adjust specific elements of the current default theme by adding a <code>theme()</code> layer and passing in arguments for the things we wish to change. Or we can use both, a built-in theme layer and a custom theme layer!</p>
<p>Let’s add a built-in theme layer <code>theme_bw()</code> first.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_ratio, <span class="at">y =</span> GO_term, , <span class="at">color =</span> p.value), </span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Aside_ggplot2_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><strong><em>Do the axis labels or the tick labels get any larger by changing themes?</em></strong></p>
<p>Not in this case. But we can add arguments using <code>theme()</code> to change it ourselves. Since we are adding this layer on top (i.e later in sequence), any features we change will override what is set in the <code>theme_bw()</code>. Here we’ll <strong>increase the size of the axes labels to be 1.15 times the default size and the x-axis tick labels to be 1.15 times the default.</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_ratio, <span class="at">y =</span> GO_term, <span class="at">color =</span> p.value), </span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>)),</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Aside_ggplot2_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>When modifying the size of text we often use the <code>rel()</code> function to specify the size we want relative to the default. We can also provide a numeric value as we did with the data point size, but it can be cumbersome if you don’t know what the default font size is to begin with.</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>You can use the <code>example("geom_point")</code> function here to explore a multitude of different aesthetics and layers that can be added to your plot. As you scroll through the different plots, take note of how the code is modified. You can use this with any of the different <code>geom</code> layers available in <code>ggplot2</code> to learn how you can easily modify your plots!</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>RStudio provides this very <a href="https://rstudio.github.io/cheatsheets/html/data-visualization.html">useful cheatsheet</a> for plotting using <code>ggplot2</code>. Different example plots are provided and the associated code (i.e which <code>geom</code> or <code>theme</code> to use in the appropriate situation.)</p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercises
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li>The current axis label text defaults to what we gave as input to <code>geom_point</code> (i.e the column headers). We can change this by <strong>adding additional layers</strong> called <code>xlab()</code> and <code>ylab()</code> for the x- and y-axis, respectively. Add these layers to the current plot such that:
<ul>
<li><strong>x-axis label:</strong> “Gene ratios”</li>
<li><strong>y-axis label:</strong> “Top 30 significant GO terms”</li>
</ul></li>
<li>Add a <code>ggtitle()</code> layer to add a title to your plot.</li>
</ol>
<p><em>NOTE: Useful code to center your title over your plot can be done using <code>theme(plot.title=element_text(hjust=0.5, face = "bold"))</code>.</em></p>
</div>
</div>
<section id="customizing-data-point-colors" class="level3">
<h3 class="anchored" data-anchor-id="customizing-data-point-colors">Customizing data point colors</h3>
<p>The plot is looking better, but it is hard to distinguish differences in significance based on the colors used. There are cheatsheets available for specifying the base R colors by <a href="https://greggilbertlab.sites.ucsc.edu/teaching/rtransition/">name</a> or <a href="http://www.r-graph-gallery.com/41-value-of-the-col-function/">hexadecimal</a> code. We could specify other colors available or use pre-created color palettes from an external R package.</p>
<p>To make additional color palettes available for plotting, we can load the RColorBrewer library, which contains color palettes designed specifically for the different types of data being compared.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Install the RColorBrewer library (if you don't have it installed already)</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"RColorBrewer"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the RColorBrewer library</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Check the available color palettes</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="fu">display.brewer.all</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Aside_ggplot2_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The output is separated into three sections based on the suggested palettes for sequential, qualitative, and diverging data.</p>
<ul>
<li><strong>Sequential palettes (top):</strong> For sequential data, with lighter colors for low values and darker colors for high values.</li>
<li><strong>Qualitative palettes (middle):</strong> For categorical data, where the color does not denote differences in magnitude or value.</li>
<li><strong>Diverging palettes (bottom):</strong> For data with emphasis on mid-range values and extremes.</li>
</ul>
<p>Since our adjusted p-values are sequential, we will choose from these palettes. Let’s go with the “Yellow, orange, red” palette. We can choose how many colors from the palette to include, which may take some trial and error. We can test the colors included in a palette by using the <code>display.brewer.pal()</code> function, and changing if desired:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Testing the palette with six colors</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">display.brewer.pal</span>(<span class="dv">6</span>, <span class="st">"YlOrRd"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Aside_ggplot2_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The yellow might be a bit too light, and we might not need so many different colors. Let’s test with three different colors:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Testing the palette with three colors</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">display.brewer.pal</span>(<span class="dv">3</span>, <span class="st">"YlOrRd"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Aside_ggplot2_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define a palette</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>mypalette <span class="ot">&lt;-</span> <span class="fu">brewer.pal</span>(<span class="dv">3</span>, <span class="st">"YlOrRd"</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="co"># how are the colors represented in the mypalette vector?</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>mypalette</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "#FFEDA0" "#FEB24C" "#F03B20"</code></pre>
</div>
</div>
<p>Those colors look okay, so let’s test them in our plot. We can add a color scale layer, and most often one of the following two scales will work:</p>
<ul>
<li><strong><code>scale_color_manual()</code>:</strong> for categorical data or quantiles</li>
<li><strong><code>scale_color_gradient()</code> family:</strong> for continuous data.</li>
</ul>
<p>By default, <code>scale_color_gradient()</code> creates a two color gradient from low to high. Since we plan to use more colors, we will use the more flexible <code>scale_color_gradientn()</code> function. To make the legend a bit cleaner, we will also perform a -log10 transform on the p-values (higher values means more significant).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_ratio, <span class="at">y =</span> GO_term, <span class="at">color =</span> <span class="sc">-</span><span class="fu">log10</span>(p.value)), </span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>)),</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>))) <span class="sc">+</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Gene ratios"</span>) <span class="sc">+</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Top 30 significant GO terms"</span>) <span class="sc">+</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Dotplot of top 30 significant GO terms"</span>) <span class="sc">+</span></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust=</span><span class="fl">0.5</span>, </span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">face =</span> <span class="st">"bold"</span>)) <span class="sc">+</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_gradientn</span>(<span class="at">colors =</span> mypalette)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Aside_ggplot2_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>This looks good, but we want to add better name for the legend and we want to make sure the legend title is centered and bold. To do this, we can add a <code>name</code> argument to <code>scale_color_gradientn()</code> and a new theme layer for the legend title.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_ratio, <span class="at">y =</span> GO_term, <span class="at">color =</span> <span class="sc">-</span><span class="fu">log10</span>(p.value)), </span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>)),</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>))) <span class="sc">+</span></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Gene ratios"</span>) <span class="sc">+</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Top 30 significant GO terms"</span>) <span class="sc">+</span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Dotplot of top 30 significant GO terms"</span>) <span class="sc">+</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust=</span><span class="fl">0.5</span>, </span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">face =</span> <span class="st">"bold"</span>)) <span class="sc">+</span></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_gradientn</span>(<span class="at">name =</span> <span class="st">"Significance </span><span class="sc">\n</span><span class="st"> (-log10(padj))"</span>, <span class="at">colors =</span> mypalette) <span class="sc">+</span></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>),</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">hjust=</span><span class="fl">0.5</span>, </span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">face=</span><span class="st">"bold"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Aside_ggplot2_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>Exercises</strong></p>
<ol type="1">
<li>Arrange <code>bp_oe</code> by <code>term_percent</code> in descending order.</li>
<li>Create a dotplot with the top 30 GO terms with highest <code>term_percent</code>, with <code>term_percent</code> as x-axis and <code>GO_term</code> as the y-axis.</li>
<li>[Optional] Color the plot using the palette of your choice.</li>
</ol>
</div>
</div>
<p>So far we have explored many layers that can be added to any plot with the ggplot2 package. However, we haven’t explored the different <code>geom</code>s available. The type of data you are plotting will determine the type of <code>geom</code> needed, but a nice summary of the main <code>geom</code>s is available on the <a href="https://rstudio.github.io/cheatsheets/data-visualization.pdf">RStudio ggplot2 cheatsheet</a>.</p>
<p>Let’s explore different <code>geom</code>s by creating a couple of different plots. We’ll start with a bar plot of the number of genes per category. We can start with the most basic plot by specifying the dataframe, geom, and aesthetics.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>(<span class="fu">aes</span>(<span class="at">x =</span> GO_term, <span class="at">y =</span> overlap.size))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Aside_ggplot2_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>This is a good base to start from, now let’s start to customize. To add color to the bars, we can use the <code>fill</code> argument, and if we would like to add an outline color to the bars, we can use the <code>color</code> argument.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>(<span class="fu">aes</span>(<span class="at">x =</span> GO_term, <span class="at">y =</span> overlap.size),</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>           <span class="at">fill =</span> <span class="st">"royalblue"</span>,</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>           <span class="at">color =</span> <span class="st">"black"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Aside_ggplot2_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Then we can provide our theme preferences, give the plot a title, and label our axes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>(<span class="fu">aes</span>(<span class="at">x =</span> GO_term, <span class="at">y =</span> overlap.size),</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>           <span class="at">fill =</span> <span class="st">"royalblue"</span>,</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>           <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>)),</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>))) <span class="sc">+</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust=</span><span class="fl">0.5</span>, </span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">face =</span> <span class="st">"bold"</span>)) <span class="sc">+</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"DE genes per GO process"</span>, <span class="at">x =</span> <span class="cn">NULL</span>, <span class="at">y =</span>  <span class="st">"# DE genes"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Aside_ggplot2_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Note that instead of using the functions <code>xlab()</code>, <code>ylab()</code>, and <code>ggtitle()</code>, we can provide all as arguments to the <code>labs()</code> function.</p>
<p>Now we might be fairly happy with our plot, but the x-axis labelling needs some help. Within the <code>theme()</code> layer, we can change the orientiation of the x-axis labels with the <code>angle</code> argument and align the labels to the x-axis with the <code>hjust</code> argument.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>(<span class="fu">aes</span>(<span class="at">x =</span> GO_term, <span class="at">y =</span> overlap.size),</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>           <span class="at">fill =</span> <span class="st">"royalblue"</span>,</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>           <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>)),</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>))) <span class="sc">+</span></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust=</span><span class="fl">0.5</span>, </span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">face =</span> <span class="st">"bold"</span>)) <span class="sc">+</span></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"DE genes per GO process"</span>, <span class="at">x =</span> <span class="cn">NULL</span>, <span class="at">y =</span>  <span class="st">"# DE genes"</span>) <span class="sc">+</span></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Aside_ggplot2_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>This is almost what we were looking for, but the labels are getting cut-off because the plotting area is too small. The <code>plot.margin</code> argument of the theme’s <code>element_text()</code> function can be used to alter the plotting dimensions to make room for our labels.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>(<span class="fu">aes</span>(<span class="at">x =</span> GO_term, <span class="at">y =</span> overlap.size),</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>           <span class="at">fill =</span> <span class="st">"royalblue"</span>,</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>           <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>)),</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>))) <span class="sc">+</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust=</span><span class="fl">0.5</span>, </span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">face =</span> <span class="st">"bold"</span>)) <span class="sc">+</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"DE genes per GO process"</span>, <span class="at">x =</span> <span class="cn">NULL</span>, <span class="at">y =</span>  <span class="st">"# DE genes"</span>) <span class="sc">+</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, </span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">hjust =</span> <span class="dv">1</span>)) <span class="sc">+</span> </span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.margin =</span> <span class="fu">unit</span>(<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">3</span>), <span class="st">"cm"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Aside_ggplot2_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>If we wanted to remove the space between the x-axis and the labels, we could add an additional layer for <code>scale_y_continuous(expand = c(0, 0))</code>, which would not expand the y-axis past the plotting limits.</p>
</div>
</div>
</section>
<section id="exporting-figures-to-file" class="level3">
<h3 class="anchored" data-anchor-id="exporting-figures-to-file">Exporting figures to file</h3>
<p>There are two ways in which figures and plots can be output to a file (rather than simply displaying on screen). The first (and easiest) is to export directly from the RStudio ‘Plots’ panel, by clicking on <code>Export</code> when the image is plotted. This will give you the option of <code>png</code> or <code>pdf</code> and selecting the directory to which you wish to save it to. It will also give you options to dictate the size and resolution of the output image.</p>
<p>The second option is to use R functions and have the write to file hard-coded in to your script. This would allow you to run the script from start to finish and automate the process (not requiring human point-and-click actions to save).</p>
</section>
<section id="consistent-formatting-using-custom-functions" class="level3">
<h3 class="anchored" data-anchor-id="consistent-formatting-using-custom-functions">Consistent formatting using custom functions</h3>
<p>When publishing, it is helpful to ensure all plots have similar formatting. To do this we can create a custom function with our preferences for the theme.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="do">## DO NOT RUN</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>name_of_function <span class="ot">&lt;-</span> <span class="cf">function</span>(arguments) {</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>    statements or code that does something</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, let’s suppose we always wanted our theme to include the following:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="do">## DO NOT RUN</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>)),</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>)),</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">10</span>, </span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">face=</span><span class="st">"bold"</span>),</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">plot.title=</span><span class="fu">element_text</span>(<span class="at">hjust=</span><span class="fl">0.5</span>, </span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>                                <span class="at">face =</span> <span class="st">"bold"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If there is nothing that we want to change when we run this, then we do not need to specify any arguments. Creating the function is simple; we can just put the code inside the <code>{}</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>personal_theme <span class="ot">&lt;-</span> <span class="cf">function</span>(){</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>)),</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>)),</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">10</span>, </span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">face=</span><span class="st">"bold"</span>),</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">plot.title=</span><span class="fu">element_text</span>(<span class="at">hjust=</span><span class="fl">0.5</span>, </span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>                                <span class="at">face =</span> <span class="st">"bold"</span>))</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now to run our personal theme with any plot, we can use this function in place of all of the <code>theme()</code> layers:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_ratio, <span class="at">y =</span> GO_term, <span class="at">color =</span> p.value), </span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">personal_theme</span>() <span class="sc">+</span></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Gene ratios"</span>) <span class="sc">+</span></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Top 30 significant GO terms"</span>) <span class="sc">+</span></span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Dotplot of top 30 significant GO terms"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Aside_ggplot2_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Based on the number of genes associated with each GO term (“term.size” column) we can categorize them into “small”, “large” or “medium” categories. Once we have done that, we want to determine what the spread of p-values is for each category; we can do this by drawing a boxplot.</p>
<ol type="1">
<li>Use the following code to create a new column in <code>bp_oe</code> tibble for the new categories.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> bp_oe<span class="sc">$</span>term.size</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>    sizes <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA</span>, <span class="fu">length</span>(x) )</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>    sizes[<span class="fu">which</span>(x <span class="sc">&gt;</span> <span class="dv">3000</span>)] <span class="ot">&lt;-</span> <span class="st">"large"</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>    sizes[<span class="fu">which</span>(x <span class="sc">&lt;=</span> <span class="dv">3000</span> <span class="sc">&amp;</span> x <span class="sc">&gt;</span> <span class="dv">500</span> )] <span class="ot">&lt;-</span> <span class="st">"medium"</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>    sizes[<span class="fu">which</span>(x <span class="sc">&lt;=</span> <span class="dv">500</span>)] <span class="ot">&lt;-</span> <span class="st">"small"</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>    bp_oe<span class="sc">$</span>term_cat <span class="ot">&lt;-</span> <span class="fu">factor</span>(sizes, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"small"</span>,<span class="st">"medium"</span>,<span class="st">"large"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="1">
<li><p>Create a boxplot with the new column (<code>term.cat</code>) on the x-axis and the -log10 of the p.value on the y-axis.</p></li>
<li><p>Fill color into each boxplot based on that new column</p></li>
<li><p>Add appropriate labels and <code>theme()</code> layers to your liking.</p></li>
</ol>
</div>
</div>
</section>
</section>
<section id="resources" class="level2">
<h2 class="anchored" data-anchor-id="resources">Resources</h2>
<p>Helpful packages to add additional functionality to ggplot2:</p>
<ul>
<li><a href="https://cran.r-project.org/web/packages/cowplot/vignettes/introduction.html">cowplot</a></li>
<li><a href="https://rpkgs.datanovia.com/ggpubr/index.html">ggpubr</a></li>
<li><a href="https://medium.com/bbc-visual-and-data-journalism/how-the-bbc-visual-and-data-journalism-team-works-with-graphics-in-r-ed0b35693535">bbplot</a></li>
<li><a href="https://cran.r-project.org/web/packages/ggrepel/vignettes/ggrepel.html">ggrepel</a></li>
<li><a href="https://serialmentor.com/dataviz/">Fundamentals of Data Visualization</a> by <a href="https://github.com/clauswilke">Claus 0. Wilke</a></li>
</ul>


<!-- -->

</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb31" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> Plotting and data visualization in R</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> Mary Piper, Meeta Mistry, Radhika Khetani</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> "Wednesday, August 23rd, 2023"</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>Approximate time: 90 minutes</span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a><span class="fu">## Learning Objectives </span></span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Explain the grammar of graphics syntax used by ggplot2</span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Determine how to plot different types of graphs with ggplot2 depending on the number and type of variables</span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Export plots for use outside of the R environment.</span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a><span class="fu">## Data Visualization with `ggplot2`</span></span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-17"><a href="#cb31-17" aria-hidden="true" tabindex="-1"></a>When we are working with large sets of numbers it can be useful to display that information graphically to gain more insight. Visualization deserves an entire course of its own (there is that much to know!). In this lesson we will be plotting with the popular package <span class="co">[</span><span class="ot">`ggplot2`</span><span class="co">](http://docs.ggplot2.org/)</span>.</span>
<span id="cb31-18"><a href="#cb31-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-19"><a href="#cb31-19" aria-hidden="true" tabindex="-1"></a>More recently, R users have moved away from base graphic options towards <span class="in">`ggplot2`</span> since it offers a lot more functionality as compared to the base R plotting functions. The <span class="in">`ggplot2`</span> syntax takes some getting used to, but once you get it, you will find it's extremely powerful and flexible. We will start with drawing a simple X-Y scatterplot of <span class="in">`gene_ratio`</span> versus <span class="in">`GO_term`</span> from the <span class="in">`bp_oe`</span> tibble. <span class="in">`ggplot2`</span> expects that the information being plotted is contained in a data frame or tibble (data frame-like).</span>
<span id="cb31-20"><a href="#cb31-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-21"><a href="#cb31-21" aria-hidden="true" tabindex="-1"></a>We would typically start by loading the <span class="in">`ggplot2`</span> library, but it is a part of the <span class="in">`tidyverse`</span> suite, so it was loaded in the last lesson.</span>
<span id="cb31-22"><a href="#cb31-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-25"><a href="#cb31-25" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-26"><a href="#cb31-26" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"data/GOs_for_plotting.RData"</span>)</span>
<span id="cb31-27"><a href="#cb31-27" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-28"><a href="#cb31-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-29"><a href="#cb31-29" aria-hidden="true" tabindex="-1"></a>**ggplot2 syntax:** To **initialize the basic graph structure** with this package we have to use the <span class="in">`ggplot()`</span> function, then we add "layers" to it using the <span class="in">`+`</span> operator. The idea is that you create a basic plot first, then additional functions are added on to build the final plot.</span>
<span id="cb31-30"><a href="#cb31-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-31"><a href="#cb31-31" aria-hidden="true" tabindex="-1"></a>Let's start by first subsetting the <span class="in">`bp_oe`</span> dataframe to only contain the top 30 most significant GO terms: </span>
<span id="cb31-32"><a href="#cb31-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-35"><a href="#cb31-35" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-36"><a href="#cb31-36" aria-hidden="true" tabindex="-1"></a><span class="co"># Load ggplot2 package</span></span>
<span id="cb31-37"><a href="#cb31-37" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb31-38"><a href="#cb31-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-39"><a href="#cb31-39" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset data frame</span></span>
<span id="cb31-40"><a href="#cb31-40" aria-hidden="true" tabindex="-1"></a>bp_plot <span class="ot">&lt;-</span> bp_oe[<span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, ]</span>
<span id="cb31-41"><a href="#cb31-41" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-42"><a href="#cb31-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-43"><a href="#cb31-43" aria-hidden="true" tabindex="-1"></a>Now let's try making a plot:</span>
<span id="cb31-44"><a href="#cb31-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-47"><a href="#cb31-47" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-48"><a href="#cb31-48" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb31-49"><a href="#cb31-49" aria-hidden="true" tabindex="-1"></a><span class="co"># What happens if we don't provide it a geom object?</span></span>
<span id="cb31-50"><a href="#cb31-50" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot)</span>
<span id="cb31-51"><a href="#cb31-51" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-52"><a href="#cb31-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-53"><a href="#cb31-53" aria-hidden="true" tabindex="-1"></a>You get a blank plot, because you need to **specify layers** using the <span class="in">`+`</span> operator.</span>
<span id="cb31-54"><a href="#cb31-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-55"><a href="#cb31-55" aria-hidden="true" tabindex="-1"></a>One type of layer is **geometric objects**. This is a mandatory layer and it specifies what type of plot you are interested in making. Examples include:</span>
<span id="cb31-56"><a href="#cb31-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-57"><a href="#cb31-57" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>points (<span class="in">`geom_point()`</span>, <span class="in">`geom_jitter()`</span> for scatter plots, dot plots, etc)</span>
<span id="cb31-58"><a href="#cb31-58" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>lines (<span class="in">`geom_line()`</span>, for time series, trend lines, etc)</span>
<span id="cb31-59"><a href="#cb31-59" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>boxplot (<span class="in">`geom_boxplot()`</span>, for, well, boxplots!)</span>
<span id="cb31-60"><a href="#cb31-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-61"><a href="#cb31-61" aria-hidden="true" tabindex="-1"></a>For a more exhaustive list on all possible geometric objects and when to use them check out <span class="co">[</span><span class="ot">Hadley Wickham's RPubs</span><span class="co">](http://rpubs.com/hadley/ggplot2-layers)</span> or the <span class="co">[</span><span class="ot">RStudio cheatsheet</span><span class="co">](https://rstudio.github.io/cheatsheets/html/data-visualization.html)</span>. </span>
<span id="cb31-62"><a href="#cb31-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-63"><a href="#cb31-63" aria-hidden="true" tabindex="-1"></a>A plot **must have at least one `geom`**, and there can be multiple *complementary* <span class="in">`geom`</span>s; there is no upper limit. </span>
<span id="cb31-64"><a href="#cb31-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-65"><a href="#cb31-65" aria-hidden="true" tabindex="-1"></a>Let's add a <span class="in">`geom`</span> to make a scatter plot, i.e. the <span class="in">`geom_point()`</span> function</span>
<span id="cb31-66"><a href="#cb31-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-69"><a href="#cb31-69" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-70"><a href="#cb31-70" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb31-71"><a href="#cb31-71" aria-hidden="true" tabindex="-1"></a><span class="co"># Provide the plot a dataset and a type of plot</span></span>
<span id="cb31-72"><a href="#cb31-72" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb31-73"><a href="#cb31-73" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() </span>
<span id="cb31-74"><a href="#cb31-74" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-75"><a href="#cb31-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-76"><a href="#cb31-76" aria-hidden="true" tabindex="-1"></a>You will find that even though we have added a layer by specifying <span class="in">`geom_point`</span>, we get an error. This is because each type of geom usually has a required set of **aesthetics**. Aesthetic mappings are set with the <span class="in">`aes()`</span> function which can be nested within the <span class="in">`geom`</span> function and can be set inside <span class="in">`geom_point()`</span> to be specifically applied to that layer. If we supplied aesthetics within <span class="in">`ggplot()`</span>, they will be used as defaults for every layer. Below are some examples of what is categorized as aesthetics in the ggplot2 context:</span>
<span id="cb31-77"><a href="#cb31-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-78"><a href="#cb31-78" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>position (i.e. columns to be used for the x and y axes)</span>
<span id="cb31-79"><a href="#cb31-79" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>color ("outside" color of the data point/bar being plotted)</span>
<span id="cb31-80"><a href="#cb31-80" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>fill ("inside" color of the data point/bar being plotted) </span>
<span id="cb31-81"><a href="#cb31-81" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>shape (of data points)</span>
<span id="cb31-82"><a href="#cb31-82" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>linetype</span>
<span id="cb31-83"><a href="#cb31-83" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>size (of data points)</span>
<span id="cb31-84"><a href="#cb31-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-85"><a href="#cb31-85" aria-hidden="true" tabindex="-1"></a>To start, we will specify the columns for the x- and y-axis since <span class="in">`geom_point()`</span> requires the most basic information about a scatterplot, i.e. what you want to plot on the x and y axes.</span>
<span id="cb31-86"><a href="#cb31-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-87"><a href="#cb31-87" aria-hidden="true" tabindex="-1"></a>Typically, a scatterplot is used to illustrate the relationship between two numeric variables. The x-axis represents the independent variable and the y-axis represents the dependent variable. We can test this out using two numeric columns from our <span class="in">`bp_plot`</span> tibble:</span>
<span id="cb31-88"><a href="#cb31-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-91"><a href="#cb31-91" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-92"><a href="#cb31-92" aria-hidden="true" tabindex="-1"></a><span class="co"># Provide the plot a dataset, a type of plot and an aesthetics function</span></span>
<span id="cb31-93"><a href="#cb31-93" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb31-94"><a href="#cb31-94" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> overlap.size, <span class="at">y =</span> p.value))</span>
<span id="cb31-95"><a href="#cb31-95" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-96"><a href="#cb31-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-97"><a href="#cb31-97" aria-hidden="true" tabindex="-1"></a>However, instead of a scatterplot with numeric values on both axes, we would like to create a dotplot for visualizing the top 30 functional categories in our dataset, and how prevalent they are. Basically, we want a dotplot for visualizing functional analysis data, which plots the gene ratio values on the x-axis and the GO terms on the y-axis.</span>
<span id="cb31-98"><a href="#cb31-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-99"><a href="#cb31-99" aria-hidden="true" tabindex="-1"></a>Let's see what happens when we add a non-numeric value to the y-axis and change the x-axis to the "gene_ratio" column:</span>
<span id="cb31-100"><a href="#cb31-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-103"><a href="#cb31-103" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-104"><a href="#cb31-104" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb31-105"><a href="#cb31-105" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_ratio, <span class="at">y =</span> GO_term))</span>
<span id="cb31-106"><a href="#cb31-106" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-107"><a href="#cb31-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-108"><a href="#cb31-108" aria-hidden="true" tabindex="-1"></a>Now that we have the required aesthetics, let's add some extras like color to the plot. Let's say we wanted *to quickly visualize significance of the GO terms* in the plot, we can **`color` the points on the plot based on p-values**, by specifying the column header.</span>
<span id="cb31-109"><a href="#cb31-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-112"><a href="#cb31-112" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-113"><a href="#cb31-113" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb31-114"><a href="#cb31-114" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_ratio, <span class="at">y =</span> GO_term, <span class="at">color =</span> p.value))</span>
<span id="cb31-115"><a href="#cb31-115" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-116"><a href="#cb31-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-117"><a href="#cb31-117" aria-hidden="true" tabindex="-1"></a>You will notice that there are a default set of colors that will be used so we do not have to specify which colors to use. Also, the **legend has been conveniently plotted for us!**</span>
<span id="cb31-118"><a href="#cb31-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-119"><a href="#cb31-119" aria-hidden="true" tabindex="-1"></a>Alternatively, we could color number of DE genes associated with each term (<span class="in">`overlap.size`</span>). </span>
<span id="cb31-120"><a href="#cb31-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-123"><a href="#cb31-123" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-124"><a href="#cb31-124" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb31-125"><a href="#cb31-125" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_ratio, <span class="at">y =</span> GO_term, <span class="at">color =</span> overlap.size))</span>
<span id="cb31-126"><a href="#cb31-126" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-127"><a href="#cb31-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-128"><a href="#cb31-128" aria-hidden="true" tabindex="-1"></a>Moving forward, we are going to stick with coloring the dots based on the p.value column. Let's explore some of the other arguments that can be specified in the <span class="in">`geom`</span> layer.</span>
<span id="cb31-129"><a href="#cb31-129" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-130"><a href="#cb31-130" aria-hidden="true" tabindex="-1"></a>To modify the **size of the data points** we can use the <span class="in">`size`</span> argument. </span>
<span id="cb31-131"><a href="#cb31-131" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>If we add <span class="in">`size`</span> inside <span class="in">`aes()`</span> we could assign a numeric column to it and the size of the data points would change according to that column. </span>
<span id="cb31-132"><a href="#cb31-132" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>However, if we add <span class="in">`size`</span> inside the <span class="in">`geom_point()`</span> but outside <span class="in">`aes()`</span> we can't assign a column to it, instead we have to give it a numeric value. This use of <span class="in">`size`</span> will uniformly change the size of all the data points.</span>
<span id="cb31-133"><a href="#cb31-133" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-134"><a href="#cb31-134" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb31-135"><a href="#cb31-135" aria-hidden="true" tabindex="-1"></a>This is true for several arguments, including <span class="in">`color`</span>, <span class="in">`shape`</span> etc. E.g. we can change all shapes to square by adding this argument to be outside the <span class="in">`aes()`</span> function; if we put the argument inside the <span class="in">`aes()`</span> function we could change the shape according to a (categorical) variable in our data frame or tibble.</span>
<span id="cb31-136"><a href="#cb31-136" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb31-137"><a href="#cb31-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-138"><a href="#cb31-138" aria-hidden="true" tabindex="-1"></a>We have decided that we want to change the size of all the data point to a uniform size instead of typing it to a numeric column in the input tibble. Add in the <span class="in">`size`</span> argument by specifying a number for the size of the data point:</span>
<span id="cb31-139"><a href="#cb31-139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-142"><a href="#cb31-142" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-143"><a href="#cb31-143" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb31-144"><a href="#cb31-144" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_ratio, <span class="at">y =</span> GO_term, , <span class="at">color =</span> p.value), </span>
<span id="cb31-145"><a href="#cb31-145" aria-hidden="true" tabindex="-1"></a>             <span class="at">size =</span> <span class="dv">2</span>)</span>
<span id="cb31-146"><a href="#cb31-146" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-147"><a href="#cb31-147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-148"><a href="#cb31-148" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb31-149"><a href="#cb31-149" aria-hidden="true" tabindex="-1"></a>The size of the points is personal preference, and you may need to play around with the parameter to decide which size is best. That seems a bit too small, so we can try out a slightly larger size. </span>
<span id="cb31-150"><a href="#cb31-150" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb31-151"><a href="#cb31-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-152"><a href="#cb31-152" aria-hidden="true" tabindex="-1"></a>As we do that, let's see how we can change the shape of the data point. Different shapes are available, as detailed in the <span class="co">[</span><span class="ot">RStudio ggplot2 cheatsheet</span><span class="co">](https://rstudio.github.io/cheatsheets/html/data-visualization.html)</span>. Let's explore this parameter by changing all of the points to squares:</span>
<span id="cb31-153"><a href="#cb31-153" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-156"><a href="#cb31-156" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-157"><a href="#cb31-157" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb31-158"><a href="#cb31-158" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_ratio, <span class="at">y =</span> GO_term, , <span class="at">color =</span> p.value), </span>
<span id="cb31-159"><a href="#cb31-159" aria-hidden="true" tabindex="-1"></a>             <span class="at">size =</span> <span class="dv">2</span>, </span>
<span id="cb31-160"><a href="#cb31-160" aria-hidden="true" tabindex="-1"></a>             <span class="at">shape =</span> <span class="st">"square"</span>)</span>
<span id="cb31-161"><a href="#cb31-161" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-162"><a href="#cb31-162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-163"><a href="#cb31-163" aria-hidden="true" tabindex="-1"></a>Now we can start updating the plot to suit our preferences for how we want the data displayed. The labels on the x- and y-axis are also quite small and not very descriptive. To change their size and labeling, we need to add additional **theme layers**. The ggplot2 <span class="in">`theme()`</span> system handles modification of non-data plot elements such as:</span>
<span id="cb31-164"><a href="#cb31-164" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-165"><a href="#cb31-165" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Axis label aesthetics</span>
<span id="cb31-166"><a href="#cb31-166" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Plot background</span>
<span id="cb31-167"><a href="#cb31-167" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Facet label backround</span>
<span id="cb31-168"><a href="#cb31-168" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Legend appearance</span>
<span id="cb31-169"><a href="#cb31-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-170"><a href="#cb31-170" aria-hidden="true" tabindex="-1"></a>There are built-in themes that we can use (i.e. <span class="in">`theme_bw()`</span>) that mostly change the background/foreground colors, by adding it as additional layer. Alternatively, we can adjust specific elements of the current default theme by adding a <span class="in">`theme()`</span> layer and passing in arguments for the things we wish to change. Or we can use both, a built-in theme layer and a custom theme layer!</span>
<span id="cb31-171"><a href="#cb31-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-172"><a href="#cb31-172" aria-hidden="true" tabindex="-1"></a>Let's add a built-in theme layer <span class="in">`theme_bw()`</span> first. </span>
<span id="cb31-173"><a href="#cb31-173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-176"><a href="#cb31-176" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-177"><a href="#cb31-177" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb31-178"><a href="#cb31-178" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_ratio, <span class="at">y =</span> GO_term, , <span class="at">color =</span> p.value), </span>
<span id="cb31-179"><a href="#cb31-179" aria-hidden="true" tabindex="-1"></a>             <span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb31-180"><a href="#cb31-180" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span>
<span id="cb31-181"><a href="#cb31-181" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-182"><a href="#cb31-182" aria-hidden="true" tabindex="-1"></a>***Do the axis labels or the tick labels get any larger by changing themes?***</span>
<span id="cb31-183"><a href="#cb31-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-184"><a href="#cb31-184" aria-hidden="true" tabindex="-1"></a>Not in this case. But we can add arguments using <span class="in">`theme()`</span> to change it ourselves. Since we are adding this layer on top (i.e later in sequence), any features we change will override what is set in the <span class="in">`theme_bw()`</span>. Here we'll **increase the size of the axes labels to be 1.15 times the default size and the x-axis tick labels to be 1.15 times the default.** </span>
<span id="cb31-185"><a href="#cb31-185" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-188"><a href="#cb31-188" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-189"><a href="#cb31-189" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb31-190"><a href="#cb31-190" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_ratio, <span class="at">y =</span> GO_term, <span class="at">color =</span> p.value), </span>
<span id="cb31-191"><a href="#cb31-191" aria-hidden="true" tabindex="-1"></a>             <span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb31-192"><a href="#cb31-192" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb31-193"><a href="#cb31-193" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>)),</span>
<span id="cb31-194"><a href="#cb31-194" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>)))</span>
<span id="cb31-195"><a href="#cb31-195" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-196"><a href="#cb31-196" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb31-197"><a href="#cb31-197" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb31-198"><a href="#cb31-198" aria-hidden="true" tabindex="-1"></a>When modifying the size of text we often use the <span class="in">`rel()`</span> function to specify the size we want relative to the default. We can also provide a numeric value as we did with the data point size, but it can be cumbersome if you don't know what the default font size is to begin with. </span>
<span id="cb31-199"><a href="#cb31-199" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb31-200"><a href="#cb31-200" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-201"><a href="#cb31-201" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb31-202"><a href="#cb31-202" aria-hidden="true" tabindex="-1"></a>You can use the <span class="in">`example("geom_point")`</span> function here to explore a multitude of different aesthetics and layers that can be added to your plot. As you scroll through the different plots, take note of how the code is modified. You can use this with any of the different <span class="in">`geom`</span> layers available in <span class="in">`ggplot2`</span> to learn how you can easily modify your plots! </span>
<span id="cb31-203"><a href="#cb31-203" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb31-204"><a href="#cb31-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-205"><a href="#cb31-205" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb31-206"><a href="#cb31-206" aria-hidden="true" tabindex="-1"></a>RStudio provides this very <span class="co">[</span><span class="ot">useful cheatsheet</span><span class="co">](https://rstudio.github.io/cheatsheets/html/data-visualization.html)</span> for plotting using <span class="in">`ggplot2`</span>. Different example plots are provided and the associated code (i.e which <span class="in">`geom`</span> or <span class="in">`theme`</span> to use in the appropriate situation.)</span>
<span id="cb31-207"><a href="#cb31-207" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb31-208"><a href="#cb31-208" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-209"><a href="#cb31-209" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb31-210"><a href="#cb31-210" aria-hidden="true" tabindex="-1"></a><span class="fu"># Exercises</span></span>
<span id="cb31-211"><a href="#cb31-211" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-212"><a href="#cb31-212" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>The current axis label text defaults to what we gave as input to <span class="in">`geom_point`</span> (i.e the column headers). We can change this by **adding additional layers** called <span class="in">`xlab()`</span> and <span class="in">`ylab()`</span> for the x- and y-axis, respectively. Add these layers to the current plot such that:</span>
<span id="cb31-213"><a href="#cb31-213" aria-hidden="true" tabindex="-1"></a><span class="ss">    - </span>**x-axis label:** "Gene ratios"</span>
<span id="cb31-214"><a href="#cb31-214" aria-hidden="true" tabindex="-1"></a><span class="ss">    - </span>**y-axis label:** "Top 30 significant GO terms"</span>
<span id="cb31-215"><a href="#cb31-215" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Add a <span class="in">`ggtitle()`</span> layer to add a title to your plot. </span>
<span id="cb31-216"><a href="#cb31-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-217"><a href="#cb31-217" aria-hidden="true" tabindex="-1"></a>*NOTE: Useful code to center your title over your plot can be done using `theme(plot.title=element_text(hjust=0.5, face = "bold"))`.*</span>
<span id="cb31-218"><a href="#cb31-218" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb31-219"><a href="#cb31-219" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-220"><a href="#cb31-220" aria-hidden="true" tabindex="-1"></a><span class="fu">### Customizing data point colors</span></span>
<span id="cb31-221"><a href="#cb31-221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-222"><a href="#cb31-222" aria-hidden="true" tabindex="-1"></a>The plot is looking better, but it is hard to distinguish differences in significance based on the colors used. There are cheatsheets available for specifying the base R colors by <span class="co">[</span><span class="ot">name</span><span class="co">](https://greggilbertlab.sites.ucsc.edu/teaching/rtransition/)</span> or <span class="co">[</span><span class="ot">hexadecimal</span><span class="co">](http://www.r-graph-gallery.com/41-value-of-the-col-function/)</span> code. We could specify other colors available or use pre-created color palettes from an external R package. </span>
<span id="cb31-223"><a href="#cb31-223" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-224"><a href="#cb31-224" aria-hidden="true" tabindex="-1"></a>To make additional color palettes available for plotting, we can load the RColorBrewer library, which contains color palettes designed specifically for the different types of data being compared.</span>
<span id="cb31-225"><a href="#cb31-225" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-228"><a href="#cb31-228" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-229"><a href="#cb31-229" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb31-230"><a href="#cb31-230" aria-hidden="true" tabindex="-1"></a><span class="co"># Install the RColorBrewer library (if you don't have it installed already)</span></span>
<span id="cb31-231"><a href="#cb31-231" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"RColorBrewer"</span>)</span>
<span id="cb31-232"><a href="#cb31-232" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-233"><a href="#cb31-233" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-236"><a href="#cb31-236" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-237"><a href="#cb31-237" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the RColorBrewer library</span></span>
<span id="cb31-238"><a href="#cb31-238" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb31-239"><a href="#cb31-239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-240"><a href="#cb31-240" aria-hidden="true" tabindex="-1"></a><span class="co"># Check the available color palettes</span></span>
<span id="cb31-241"><a href="#cb31-241" aria-hidden="true" tabindex="-1"></a><span class="fu">display.brewer.all</span>()</span>
<span id="cb31-242"><a href="#cb31-242" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-243"><a href="#cb31-243" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-244"><a href="#cb31-244" aria-hidden="true" tabindex="-1"></a>The output is separated into three sections based on the suggested palettes for sequential, qualitative, and diverging data. </span>
<span id="cb31-245"><a href="#cb31-245" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-246"><a href="#cb31-246" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Sequential palettes (top):** For sequential data, with lighter colors for low values and darker colors for high values.</span>
<span id="cb31-247"><a href="#cb31-247" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Qualitative palettes (middle):** For categorical data, where the color does not denote differences in magnitude or value.</span>
<span id="cb31-248"><a href="#cb31-248" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Diverging palettes (bottom):** For data with emphasis on mid-range values and extremes.</span>
<span id="cb31-249"><a href="#cb31-249" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-250"><a href="#cb31-250" aria-hidden="true" tabindex="-1"></a>Since our adjusted p-values are sequential, we will choose from these palettes. Let's go with the "Yellow, orange, red" palette. We can choose how many colors from the palette to include, which may take some trial and error. We can test the colors included in a palette by using the <span class="in">`display.brewer.pal()`</span> function, and changing if desired:</span>
<span id="cb31-251"><a href="#cb31-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-254"><a href="#cb31-254" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-255"><a href="#cb31-255" aria-hidden="true" tabindex="-1"></a><span class="co"># Testing the palette with six colors</span></span>
<span id="cb31-256"><a href="#cb31-256" aria-hidden="true" tabindex="-1"></a><span class="fu">display.brewer.pal</span>(<span class="dv">6</span>, <span class="st">"YlOrRd"</span>)</span>
<span id="cb31-257"><a href="#cb31-257" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-258"><a href="#cb31-258" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-259"><a href="#cb31-259" aria-hidden="true" tabindex="-1"></a>The yellow might be a bit too light, and we might not need so many different colors. Let's test with three different colors:</span>
<span id="cb31-260"><a href="#cb31-260" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-263"><a href="#cb31-263" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-264"><a href="#cb31-264" aria-hidden="true" tabindex="-1"></a><span class="co"># Testing the palette with three colors</span></span>
<span id="cb31-265"><a href="#cb31-265" aria-hidden="true" tabindex="-1"></a><span class="fu">display.brewer.pal</span>(<span class="dv">3</span>, <span class="st">"YlOrRd"</span>)</span>
<span id="cb31-266"><a href="#cb31-266" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-267"><a href="#cb31-267" aria-hidden="true" tabindex="-1"></a><span class="co"># Define a palette</span></span>
<span id="cb31-268"><a href="#cb31-268" aria-hidden="true" tabindex="-1"></a>mypalette <span class="ot">&lt;-</span> <span class="fu">brewer.pal</span>(<span class="dv">3</span>, <span class="st">"YlOrRd"</span>)</span>
<span id="cb31-269"><a href="#cb31-269" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-270"><a href="#cb31-270" aria-hidden="true" tabindex="-1"></a><span class="co"># how are the colors represented in the mypalette vector?</span></span>
<span id="cb31-271"><a href="#cb31-271" aria-hidden="true" tabindex="-1"></a>mypalette</span>
<span id="cb31-272"><a href="#cb31-272" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-273"><a href="#cb31-273" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-274"><a href="#cb31-274" aria-hidden="true" tabindex="-1"></a>Those colors look okay, so let's test them in our plot. We can add a color scale layer, and most often one of the following two scales will work:</span>
<span id="cb31-275"><a href="#cb31-275" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-276"><a href="#cb31-276" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**`scale_color_manual()`:** for categorical data or quantiles</span>
<span id="cb31-277"><a href="#cb31-277" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**`scale_color_gradient()` family:** for continuous data. </span>
<span id="cb31-278"><a href="#cb31-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-279"><a href="#cb31-279" aria-hidden="true" tabindex="-1"></a>By default, <span class="in">`scale_color_gradient()`</span> creates a two color gradient from low to high. Since we plan to use more colors, we will use the more flexible <span class="in">`scale_color_gradientn()`</span> function. To make the legend a bit cleaner, we will also perform a -log10 transform on the p-values (higher values means more significant).</span>
<span id="cb31-280"><a href="#cb31-280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-283"><a href="#cb31-283" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-284"><a href="#cb31-284" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb31-285"><a href="#cb31-285" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_ratio, <span class="at">y =</span> GO_term, <span class="at">color =</span> <span class="sc">-</span><span class="fu">log10</span>(p.value)), </span>
<span id="cb31-286"><a href="#cb31-286" aria-hidden="true" tabindex="-1"></a>             <span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb31-287"><a href="#cb31-287" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb31-288"><a href="#cb31-288" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>)),</span>
<span id="cb31-289"><a href="#cb31-289" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>))) <span class="sc">+</span></span>
<span id="cb31-290"><a href="#cb31-290" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Gene ratios"</span>) <span class="sc">+</span></span>
<span id="cb31-291"><a href="#cb31-291" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Top 30 significant GO terms"</span>) <span class="sc">+</span></span>
<span id="cb31-292"><a href="#cb31-292" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Dotplot of top 30 significant GO terms"</span>) <span class="sc">+</span></span>
<span id="cb31-293"><a href="#cb31-293" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust=</span><span class="fl">0.5</span>, </span>
<span id="cb31-294"><a href="#cb31-294" aria-hidden="true" tabindex="-1"></a>    <span class="at">face =</span> <span class="st">"bold"</span>)) <span class="sc">+</span></span>
<span id="cb31-295"><a href="#cb31-295" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_gradientn</span>(<span class="at">colors =</span> mypalette)</span>
<span id="cb31-296"><a href="#cb31-296" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-297"><a href="#cb31-297" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-298"><a href="#cb31-298" aria-hidden="true" tabindex="-1"></a>This looks good, but we want to add better name for the legend and we want to make sure the legend title is centered and bold. To do this, we can add a <span class="in">`name`</span> argument to <span class="in">`scale_color_gradientn()`</span> and a new theme layer for the legend title.</span>
<span id="cb31-299"><a href="#cb31-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-302"><a href="#cb31-302" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-303"><a href="#cb31-303" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb31-304"><a href="#cb31-304" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_ratio, <span class="at">y =</span> GO_term, <span class="at">color =</span> <span class="sc">-</span><span class="fu">log10</span>(p.value)), </span>
<span id="cb31-305"><a href="#cb31-305" aria-hidden="true" tabindex="-1"></a>             <span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb31-306"><a href="#cb31-306" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb31-307"><a href="#cb31-307" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>)),</span>
<span id="cb31-308"><a href="#cb31-308" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>))) <span class="sc">+</span></span>
<span id="cb31-309"><a href="#cb31-309" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Gene ratios"</span>) <span class="sc">+</span></span>
<span id="cb31-310"><a href="#cb31-310" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Top 30 significant GO terms"</span>) <span class="sc">+</span></span>
<span id="cb31-311"><a href="#cb31-311" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Dotplot of top 30 significant GO terms"</span>) <span class="sc">+</span></span>
<span id="cb31-312"><a href="#cb31-312" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust=</span><span class="fl">0.5</span>, </span>
<span id="cb31-313"><a href="#cb31-313" aria-hidden="true" tabindex="-1"></a>    <span class="at">face =</span> <span class="st">"bold"</span>)) <span class="sc">+</span></span>
<span id="cb31-314"><a href="#cb31-314" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_gradientn</span>(<span class="at">name =</span> <span class="st">"Significance </span><span class="sc">\n</span><span class="st"> (-log10(padj))"</span>, <span class="at">colors =</span> mypalette) <span class="sc">+</span></span>
<span id="cb31-315"><a href="#cb31-315" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>),</span>
<span id="cb31-316"><a href="#cb31-316" aria-hidden="true" tabindex="-1"></a>    <span class="at">hjust=</span><span class="fl">0.5</span>, </span>
<span id="cb31-317"><a href="#cb31-317" aria-hidden="true" tabindex="-1"></a>    <span class="at">face=</span><span class="st">"bold"</span>))</span>
<span id="cb31-318"><a href="#cb31-318" aria-hidden="true" tabindex="-1"></a>             </span>
<span id="cb31-319"><a href="#cb31-319" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-320"><a href="#cb31-320" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-321"><a href="#cb31-321" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb31-322"><a href="#cb31-322" aria-hidden="true" tabindex="-1"></a>**Exercises**</span>
<span id="cb31-323"><a href="#cb31-323" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-324"><a href="#cb31-324" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Arrange <span class="in">`bp_oe`</span> by <span class="in">`term_percent`</span> in descending order.</span>
<span id="cb31-325"><a href="#cb31-325" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Create a dotplot with the top 30 GO terms with highest <span class="in">`term_percent`</span>, with <span class="in">`term_percent`</span> as x-axis and <span class="in">`GO_term`</span> as the y-axis.</span>
<span id="cb31-326"><a href="#cb31-326" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span><span class="co">[</span><span class="ot">Optional</span><span class="co">]</span> Color the plot using the palette of your choice.</span>
<span id="cb31-327"><a href="#cb31-327" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb31-328"><a href="#cb31-328" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-329"><a href="#cb31-329" aria-hidden="true" tabindex="-1"></a>So far we have explored many layers that can be added to any plot with the ggplot2 package. However, we haven't explored the different <span class="in">`geom`</span>s available. The type of data you are plotting will determine the type of <span class="in">`geom`</span> needed, but a nice summary of the main <span class="in">`geom`</span>s is available on the <span class="co">[</span><span class="ot">RStudio ggplot2 cheatsheet</span><span class="co">](https://rstudio.github.io/cheatsheets/data-visualization.pdf)</span>.</span>
<span id="cb31-330"><a href="#cb31-330" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-331"><a href="#cb31-331" aria-hidden="true" tabindex="-1"></a>Let's explore different <span class="in">`geom`</span>s by creating a couple of different plots. We'll start with a bar plot of the number of genes per category. We can start with the most basic plot by specifying the dataframe, geom, and aesthetics. </span>
<span id="cb31-332"><a href="#cb31-332" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-335"><a href="#cb31-335" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-336"><a href="#cb31-336" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb31-337"><a href="#cb31-337" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>(<span class="fu">aes</span>(<span class="at">x =</span> GO_term, <span class="at">y =</span> overlap.size))</span>
<span id="cb31-338"><a href="#cb31-338" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-339"><a href="#cb31-339" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-340"><a href="#cb31-340" aria-hidden="true" tabindex="-1"></a>This is a good base to start from, now let's start to customize. To add color to the bars, we can use the <span class="in">`fill`</span> argument, and if we would like to add an outline color to the bars, we can use the <span class="in">`color`</span> argument.</span>
<span id="cb31-341"><a href="#cb31-341" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-344"><a href="#cb31-344" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-345"><a href="#cb31-345" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb31-346"><a href="#cb31-346" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>(<span class="fu">aes</span>(<span class="at">x =</span> GO_term, <span class="at">y =</span> overlap.size),</span>
<span id="cb31-347"><a href="#cb31-347" aria-hidden="true" tabindex="-1"></a>           <span class="at">fill =</span> <span class="st">"royalblue"</span>,</span>
<span id="cb31-348"><a href="#cb31-348" aria-hidden="true" tabindex="-1"></a>           <span class="at">color =</span> <span class="st">"black"</span>)</span>
<span id="cb31-349"><a href="#cb31-349" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-350"><a href="#cb31-350" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-351"><a href="#cb31-351" aria-hidden="true" tabindex="-1"></a>Then we can provide our theme preferences, give the plot a title, and label our axes:</span>
<span id="cb31-352"><a href="#cb31-352" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-355"><a href="#cb31-355" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-356"><a href="#cb31-356" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb31-357"><a href="#cb31-357" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>(<span class="fu">aes</span>(<span class="at">x =</span> GO_term, <span class="at">y =</span> overlap.size),</span>
<span id="cb31-358"><a href="#cb31-358" aria-hidden="true" tabindex="-1"></a>           <span class="at">fill =</span> <span class="st">"royalblue"</span>,</span>
<span id="cb31-359"><a href="#cb31-359" aria-hidden="true" tabindex="-1"></a>           <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb31-360"><a href="#cb31-360" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>)),</span>
<span id="cb31-361"><a href="#cb31-361" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>))) <span class="sc">+</span></span>
<span id="cb31-362"><a href="#cb31-362" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust=</span><span class="fl">0.5</span>, </span>
<span id="cb31-363"><a href="#cb31-363" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">face =</span> <span class="st">"bold"</span>)) <span class="sc">+</span></span>
<span id="cb31-364"><a href="#cb31-364" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"DE genes per GO process"</span>, <span class="at">x =</span> <span class="cn">NULL</span>, <span class="at">y =</span>  <span class="st">"# DE genes"</span>)</span>
<span id="cb31-365"><a href="#cb31-365" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-366"><a href="#cb31-366" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-367"><a href="#cb31-367" aria-hidden="true" tabindex="-1"></a>Note that instead of using the functions <span class="in">`xlab()`</span>, <span class="in">`ylab()`</span>, and <span class="in">`ggtitle()`</span>, we can provide all as arguments to the <span class="in">`labs()`</span> function.</span>
<span id="cb31-368"><a href="#cb31-368" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-369"><a href="#cb31-369" aria-hidden="true" tabindex="-1"></a>Now we might be fairly happy with our plot, but the x-axis labelling needs some help. Within the <span class="in">`theme()`</span> layer, we can change the orientiation of the x-axis labels with the <span class="in">`angle`</span> argument and align the labels to the x-axis with the <span class="in">`hjust`</span> argument.</span>
<span id="cb31-370"><a href="#cb31-370" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-373"><a href="#cb31-373" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-374"><a href="#cb31-374" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb31-375"><a href="#cb31-375" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>(<span class="fu">aes</span>(<span class="at">x =</span> GO_term, <span class="at">y =</span> overlap.size),</span>
<span id="cb31-376"><a href="#cb31-376" aria-hidden="true" tabindex="-1"></a>           <span class="at">fill =</span> <span class="st">"royalblue"</span>,</span>
<span id="cb31-377"><a href="#cb31-377" aria-hidden="true" tabindex="-1"></a>           <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb31-378"><a href="#cb31-378" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>)),</span>
<span id="cb31-379"><a href="#cb31-379" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>))) <span class="sc">+</span></span>
<span id="cb31-380"><a href="#cb31-380" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust=</span><span class="fl">0.5</span>, </span>
<span id="cb31-381"><a href="#cb31-381" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">face =</span> <span class="st">"bold"</span>)) <span class="sc">+</span></span>
<span id="cb31-382"><a href="#cb31-382" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"DE genes per GO process"</span>, <span class="at">x =</span> <span class="cn">NULL</span>, <span class="at">y =</span>  <span class="st">"# DE genes"</span>) <span class="sc">+</span></span>
<span id="cb31-383"><a href="#cb31-383" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>))</span>
<span id="cb31-384"><a href="#cb31-384" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-385"><a href="#cb31-385" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-386"><a href="#cb31-386" aria-hidden="true" tabindex="-1"></a>This is almost what we were looking for, but the labels are getting cut-off because the plotting area is too small. The <span class="in">`plot.margin`</span> argument of the theme's <span class="in">`element_text()`</span> function can be used to alter the plotting dimensions to make room for our labels.</span>
<span id="cb31-387"><a href="#cb31-387" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-390"><a href="#cb31-390" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-391"><a href="#cb31-391" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb31-392"><a href="#cb31-392" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>(<span class="fu">aes</span>(<span class="at">x =</span> GO_term, <span class="at">y =</span> overlap.size),</span>
<span id="cb31-393"><a href="#cb31-393" aria-hidden="true" tabindex="-1"></a>           <span class="at">fill =</span> <span class="st">"royalblue"</span>,</span>
<span id="cb31-394"><a href="#cb31-394" aria-hidden="true" tabindex="-1"></a>           <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb31-395"><a href="#cb31-395" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>)),</span>
<span id="cb31-396"><a href="#cb31-396" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>))) <span class="sc">+</span></span>
<span id="cb31-397"><a href="#cb31-397" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust=</span><span class="fl">0.5</span>, </span>
<span id="cb31-398"><a href="#cb31-398" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">face =</span> <span class="st">"bold"</span>)) <span class="sc">+</span></span>
<span id="cb31-399"><a href="#cb31-399" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"DE genes per GO process"</span>, <span class="at">x =</span> <span class="cn">NULL</span>, <span class="at">y =</span>  <span class="st">"# DE genes"</span>) <span class="sc">+</span></span>
<span id="cb31-400"><a href="#cb31-400" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, </span>
<span id="cb31-401"><a href="#cb31-401" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">hjust =</span> <span class="dv">1</span>)) <span class="sc">+</span> </span>
<span id="cb31-402"><a href="#cb31-402" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.margin =</span> <span class="fu">unit</span>(<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">3</span>), <span class="st">"cm"</span>))</span>
<span id="cb31-403"><a href="#cb31-403" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-404"><a href="#cb31-404" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-405"><a href="#cb31-405" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb31-406"><a href="#cb31-406" aria-hidden="true" tabindex="-1"></a>If we wanted to remove the space between the x-axis and the labels, we could add an additional layer for <span class="in">`scale_y_continuous(expand = c(0, 0))`</span>, which would not expand the y-axis past the plotting limits.</span>
<span id="cb31-407"><a href="#cb31-407" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb31-408"><a href="#cb31-408" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-409"><a href="#cb31-409" aria-hidden="true" tabindex="-1"></a><span class="fu">### Exporting figures to file</span></span>
<span id="cb31-410"><a href="#cb31-410" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-411"><a href="#cb31-411" aria-hidden="true" tabindex="-1"></a>There are two ways in which figures and plots can be output to a file (rather than simply displaying on screen). The first (and easiest) is to export directly from the RStudio 'Plots' panel, by clicking on <span class="in">`Export`</span> when the image is plotted. This will give you the option of <span class="in">`png`</span> or <span class="in">`pdf`</span> and selecting the directory to which you wish to save it to. It will also give you options to dictate the size and resolution of the output image.</span>
<span id="cb31-412"><a href="#cb31-412" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-413"><a href="#cb31-413" aria-hidden="true" tabindex="-1"></a>The second option is to use R functions and have the write to file hard-coded in to your script. This would allow you to run the script from start to finish and automate the process (not requiring human point-and-click actions to save).</span>
<span id="cb31-414"><a href="#cb31-414" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-415"><a href="#cb31-415" aria-hidden="true" tabindex="-1"></a><span class="fu">### Consistent formatting using custom functions</span></span>
<span id="cb31-416"><a href="#cb31-416" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-417"><a href="#cb31-417" aria-hidden="true" tabindex="-1"></a>When publishing, it is helpful to ensure all plots have similar formatting. To do this we can create a custom function with our preferences for the theme. </span>
<span id="cb31-418"><a href="#cb31-418" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-421"><a href="#cb31-421" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-422"><a href="#cb31-422" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb31-423"><a href="#cb31-423" aria-hidden="true" tabindex="-1"></a><span class="do">## DO NOT RUN</span></span>
<span id="cb31-424"><a href="#cb31-424" aria-hidden="true" tabindex="-1"></a>name_of_function <span class="ot">&lt;-</span> <span class="cf">function</span>(arguments) {</span>
<span id="cb31-425"><a href="#cb31-425" aria-hidden="true" tabindex="-1"></a>    statements or code that does something</span>
<span id="cb31-426"><a href="#cb31-426" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb31-427"><a href="#cb31-427" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-428"><a href="#cb31-428" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-429"><a href="#cb31-429" aria-hidden="true" tabindex="-1"></a>Now, let's suppose we always wanted our theme to include the following:</span>
<span id="cb31-430"><a href="#cb31-430" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-433"><a href="#cb31-433" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-434"><a href="#cb31-434" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb31-435"><a href="#cb31-435" aria-hidden="true" tabindex="-1"></a><span class="do">## DO NOT RUN</span></span>
<span id="cb31-436"><a href="#cb31-436" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb31-437"><a href="#cb31-437" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>)),</span>
<span id="cb31-438"><a href="#cb31-438" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>)),</span>
<span id="cb31-439"><a href="#cb31-439" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">10</span>, </span>
<span id="cb31-440"><a href="#cb31-440" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">face=</span><span class="st">"bold"</span>),</span>
<span id="cb31-441"><a href="#cb31-441" aria-hidden="true" tabindex="-1"></a>          <span class="at">plot.title=</span><span class="fu">element_text</span>(<span class="at">hjust=</span><span class="fl">0.5</span>, </span>
<span id="cb31-442"><a href="#cb31-442" aria-hidden="true" tabindex="-1"></a>                                <span class="at">face =</span> <span class="st">"bold"</span>))</span>
<span id="cb31-443"><a href="#cb31-443" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-444"><a href="#cb31-444" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-445"><a href="#cb31-445" aria-hidden="true" tabindex="-1"></a>If there is nothing that we want to change when we run this, then we do not need to specify any arguments. Creating the function is simple; we can just put the code inside the <span class="in">`{}`</span>:</span>
<span id="cb31-446"><a href="#cb31-446" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-449"><a href="#cb31-449" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-450"><a href="#cb31-450" aria-hidden="true" tabindex="-1"></a>personal_theme <span class="ot">&lt;-</span> <span class="cf">function</span>(){</span>
<span id="cb31-451"><a href="#cb31-451" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb31-452"><a href="#cb31-452" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>)),</span>
<span id="cb31-453"><a href="#cb31-453" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="fu">rel</span>(<span class="fl">1.15</span>)),</span>
<span id="cb31-454"><a href="#cb31-454" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">10</span>, </span>
<span id="cb31-455"><a href="#cb31-455" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">face=</span><span class="st">"bold"</span>),</span>
<span id="cb31-456"><a href="#cb31-456" aria-hidden="true" tabindex="-1"></a>          <span class="at">plot.title=</span><span class="fu">element_text</span>(<span class="at">hjust=</span><span class="fl">0.5</span>, </span>
<span id="cb31-457"><a href="#cb31-457" aria-hidden="true" tabindex="-1"></a>                                <span class="at">face =</span> <span class="st">"bold"</span>))</span>
<span id="cb31-458"><a href="#cb31-458" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb31-459"><a href="#cb31-459" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-460"><a href="#cb31-460" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-461"><a href="#cb31-461" aria-hidden="true" tabindex="-1"></a>Now to run our personal theme with any plot, we can use this function in place of all of the <span class="in">`theme()`</span> layers:</span>
<span id="cb31-462"><a href="#cb31-462" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-465"><a href="#cb31-465" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-466"><a href="#cb31-466" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bp_plot) <span class="sc">+</span></span>
<span id="cb31-467"><a href="#cb31-467" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene_ratio, <span class="at">y =</span> GO_term, <span class="at">color =</span> p.value), </span>
<span id="cb31-468"><a href="#cb31-468" aria-hidden="true" tabindex="-1"></a>             <span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb31-469"><a href="#cb31-469" aria-hidden="true" tabindex="-1"></a>  <span class="fu">personal_theme</span>() <span class="sc">+</span></span>
<span id="cb31-470"><a href="#cb31-470" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Gene ratios"</span>) <span class="sc">+</span></span>
<span id="cb31-471"><a href="#cb31-471" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Top 30 significant GO terms"</span>) <span class="sc">+</span></span>
<span id="cb31-472"><a href="#cb31-472" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Dotplot of top 30 significant GO terms"</span>)</span>
<span id="cb31-473"><a href="#cb31-473" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-474"><a href="#cb31-474" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-475"><a href="#cb31-475" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb31-476"><a href="#cb31-476" aria-hidden="true" tabindex="-1"></a><span class="fu"># Exercises</span></span>
<span id="cb31-477"><a href="#cb31-477" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-478"><a href="#cb31-478" aria-hidden="true" tabindex="-1"></a>Based on the number of genes associated with each GO term ("term.size" column) we can categorize them into "small", "large" or "medium" categories. Once we have done that, we want to determine what the spread of p-values is for each category; we can do this by drawing a boxplot.</span>
<span id="cb31-479"><a href="#cb31-479" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-480"><a href="#cb31-480" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Use the following code to create a new column in <span class="in">`bp_oe`</span> tibble for the new categories.</span>
<span id="cb31-483"><a href="#cb31-483" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-484"><a href="#cb31-484" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> bp_oe<span class="sc">$</span>term.size</span>
<span id="cb31-485"><a href="#cb31-485" aria-hidden="true" tabindex="-1"></a>    sizes <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA</span>, <span class="fu">length</span>(x) )</span>
<span id="cb31-486"><a href="#cb31-486" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb31-487"><a href="#cb31-487" aria-hidden="true" tabindex="-1"></a>    sizes[<span class="fu">which</span>(x <span class="sc">&gt;</span> <span class="dv">3000</span>)] <span class="ot">&lt;-</span> <span class="st">"large"</span></span>
<span id="cb31-488"><a href="#cb31-488" aria-hidden="true" tabindex="-1"></a>    sizes[<span class="fu">which</span>(x <span class="sc">&lt;=</span> <span class="dv">3000</span> <span class="sc">&amp;</span> x <span class="sc">&gt;</span> <span class="dv">500</span> )] <span class="ot">&lt;-</span> <span class="st">"medium"</span></span>
<span id="cb31-489"><a href="#cb31-489" aria-hidden="true" tabindex="-1"></a>    sizes[<span class="fu">which</span>(x <span class="sc">&lt;=</span> <span class="dv">500</span>)] <span class="ot">&lt;-</span> <span class="st">"small"</span></span>
<span id="cb31-490"><a href="#cb31-490" aria-hidden="true" tabindex="-1"></a>    bp_oe<span class="sc">$</span>term_cat <span class="ot">&lt;-</span> <span class="fu">factor</span>(sizes, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"small"</span>,<span class="st">"medium"</span>,<span class="st">"large"</span>))</span>
<span id="cb31-491"><a href="#cb31-491" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-492"><a href="#cb31-492" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-493"><a href="#cb31-493" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Create a boxplot with the new column (<span class="in">`term.cat`</span>) on the x-axis and the -log10 of the p.value on the y-axis.</span>
<span id="cb31-494"><a href="#cb31-494" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-495"><a href="#cb31-495" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>Fill color into each boxplot based on that new column</span>
<span id="cb31-496"><a href="#cb31-496" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-497"><a href="#cb31-497" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>Add appropriate labels and <span class="in">`theme()`</span> layers to your liking.</span>
<span id="cb31-498"><a href="#cb31-498" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb31-499"><a href="#cb31-499" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-500"><a href="#cb31-500" aria-hidden="true" tabindex="-1"></a><span class="fu">## Resources</span></span>
<span id="cb31-501"><a href="#cb31-501" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-502"><a href="#cb31-502" aria-hidden="true" tabindex="-1"></a>Helpful packages to add additional functionality to ggplot2:</span>
<span id="cb31-503"><a href="#cb31-503" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-504"><a href="#cb31-504" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="co">[</span><span class="ot">cowplot</span><span class="co">](https://cran.r-project.org/web/packages/cowplot/vignettes/introduction.html)</span></span>
<span id="cb31-505"><a href="#cb31-505" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="co">[</span><span class="ot">ggpubr</span><span class="co">](https://rpkgs.datanovia.com/ggpubr/index.html)</span></span>
<span id="cb31-506"><a href="#cb31-506" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="co">[</span><span class="ot">bbplot</span><span class="co">](https://medium.com/bbc-visual-and-data-journalism/how-the-bbc-visual-and-data-journalism-team-works-with-graphics-in-r-ed0b35693535)</span></span>
<span id="cb31-507"><a href="#cb31-507" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="co">[</span><span class="ot">ggrepel</span><span class="co">](https://cran.r-project.org/web/packages/ggrepel/vignettes/ggrepel.html)</span></span>
<span id="cb31-508"><a href="#cb31-508" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="co">[</span><span class="ot">Fundamentals of Data Visualization</span><span class="co">](https://serialmentor.com/dataviz/)</span> by <span class="co">[</span><span class="ot">Claus 0. Wilke</span><span class="co">](https://github.com/clauswilke)</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>This lesson has been developed by members of the teaching team at the <a href="http://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core (HBC)</a>. <br> These are open access materials distributed under the terms of the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license</a> (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited. <br> Some materials used are adapted from work that is Copyright © <a href="http://datacarpentry.org/">Data Carpentry</a>. All Data Carpentry instructional material is made available under the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license</a> (CC BY 4.0).</p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>