<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Introduction to R – intro_to_r_hw</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="../../css/styles.css">
</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Introduction to R</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../lessons/schedule.html" rel="" target="">
 <span class="menu-text">Schedule</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://bioinformatics.sph.harvard.edu/" rel="" target="">
 <span class="menu-text">HBC</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/hbctraining/intro_r_rmd" rel="" target="">
 <span class="menu-text">GitHub</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="mailto:hbctraining@hsph.harvard.edu" rel="" target="">
 <span class="menu-text">Contact us</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false">
 <span class="menu-text">Day 1:</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../lessons/01b_introR-R-and-RStudio.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction to R and RStudio</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false">
 <span class="menu-text">Day 1 Self-learning:</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../lessons/02_introR-syntax-and-data-structures.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">R Syntax and Data Structures</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../lessons/03_introR-functions-and-arguments.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Functions in R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../lessons/06_reading_and_data_inspection.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Reading in and inspecting data</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false">
 <span class="menu-text">Day 2</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../lessons/Day2_activities.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Day 2 Activities</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../lessons/05_introR-data-wrangling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Data subsetting with base R: vectors and factors</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false">
 <span class="menu-text">Day 2 Self-learning</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../lessons/04_introR_packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Packages and libraries</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../lessons/07_introR-data-wrangling2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Data wrangling: dataframes, matrices, and lists</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../lessons/08_identifying-matching-elements.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Advanced R, logical operators for matching</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../lessons/09_reordering-to-match-datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Advanced R, reordering to match datasets</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../lessons/10_setting_up_to_plot.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Plotting and data visualization in R</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">
 <span class="menu-text">Day 3</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../lessons/Day3_activities.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Day 3 Activities</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../lessons/11_ggplot2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Plotting and data visualization in R</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false">
 <span class="menu-text">Day 3 Self-learning</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../lessons/11b_Custom_Functions_ggplot2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Custom functions for consistent plots</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../lessons/12_boxplot_exercise.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Plotting and data visualization in R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../lessons/13_exporting_data_and_plots.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Saving data and plots to file</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../lessons/14_finding_help.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Troubleshooting and finding help</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../lessons/15_tidyverse.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Tidyverse data wrangling</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="false">
 <span class="menu-text">Day 4</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../lessons/Day4_activities.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Day 4 Activities</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction-to-r-practice" id="toc-introduction-to-r-practice" class="nav-link active" data-scroll-target="#introduction-to-r-practice">Introduction to R practice</a>
  <ul class="collapse">
  <li><a href="#creating-vectorsfactors-and-dataframes" id="toc-creating-vectorsfactors-and-dataframes" class="nav-link" data-scroll-target="#creating-vectorsfactors-and-dataframes">Creating vectors/factors and dataframes</a></li>
  <li><a href="#subsetting-vectorsfactors-and-dataframes" id="toc-subsetting-vectorsfactors-and-dataframes" class="nav-link" data-scroll-target="#subsetting-vectorsfactors-and-dataframes">Subsetting vectors/factors and dataframes</a></li>
  <li><a href="#extracting-components-from-lists" id="toc-extracting-components-from-lists" class="nav-link" data-scroll-target="#extracting-components-from-lists">Extracting components from lists</a></li>
  <li><a href="#creating-figures-with-ggplot2" id="toc-creating-figures-with-ggplot2" class="nav-link" data-scroll-target="#creating-figures-with-ggplot2">Creating figures with ggplot2</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><div class="quarto-title-block"><div class="quarto-title-tools-only"><h1></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>



<section id="introduction-to-r-practice" class="level1">
<h1>Introduction to R practice</h1>
<section id="creating-vectorsfactors-and-dataframes" class="level2">
<h2 class="anchored" data-anchor-id="creating-vectorsfactors-and-dataframes">Creating vectors/factors and dataframes</h2>
<ol type="1">
<li><p>We are performing RNA-Seq on cancer samples being treated with three different types of treatment (A, B, and P). You have 12 samples total, with 4 replicates per treatment. Write the R code you would use to construct your metadata table as described below.</p>
<ul>
<li>Create the vectors/factors for each column (Hint: you can type out each vector/factor, or if you want the process go faster try exploring the <code>rep()</code> function).</li>
<li>Put them together into a dataframe called <code>meta</code>.</li>
<li>Use the <code>rownames()</code> function to assign row names to the dataframe (Hint: you can type out the row names as a vector, or if you want the process go faster try exploring the <code>paste()</code> function).</li>
</ul>
<p>Your finished metadata table should have information for the variables <code>sex</code>, <code>stage</code>, <code>treatment</code>, and <code>myc</code> levels:</p>
<table class="table">
<thead>
<tr class="header">
<th style="text-align: center;"></th>
<th style="text-align: center;">sex</th>
<th style="text-align: center;">stage</th>
<th style="text-align: center;">treatment</th>
<th style="text-align: center;">myc</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">sample1</td>
<td style="text-align: center;">M</td>
<td style="text-align: center;">I</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">2343</td>
</tr>
<tr class="even">
<td style="text-align: center;">sample2</td>
<td style="text-align: center;">F</td>
<td style="text-align: center;">II</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">457</td>
</tr>
<tr class="odd">
<td style="text-align: center;">sample3</td>
<td style="text-align: center;">M</td>
<td style="text-align: center;">II</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">4593</td>
</tr>
<tr class="even">
<td style="text-align: center;">sample4</td>
<td style="text-align: center;">F</td>
<td style="text-align: center;">I</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">9035</td>
</tr>
<tr class="odd">
<td style="text-align: center;">sample5</td>
<td style="text-align: center;">M</td>
<td style="text-align: center;">II</td>
<td style="text-align: center;">B</td>
<td style="text-align: center;">3450</td>
</tr>
<tr class="even">
<td style="text-align: center;">sample6</td>
<td style="text-align: center;">F</td>
<td style="text-align: center;">II</td>
<td style="text-align: center;">B</td>
<td style="text-align: center;">3524</td>
</tr>
<tr class="odd">
<td style="text-align: center;">sample7</td>
<td style="text-align: center;">M</td>
<td style="text-align: center;">I</td>
<td style="text-align: center;">B</td>
<td style="text-align: center;">958</td>
</tr>
<tr class="even">
<td style="text-align: center;">sample8</td>
<td style="text-align: center;">F</td>
<td style="text-align: center;">II</td>
<td style="text-align: center;">B</td>
<td style="text-align: center;">1053</td>
</tr>
<tr class="odd">
<td style="text-align: center;">sample9</td>
<td style="text-align: center;">M</td>
<td style="text-align: center;">II</td>
<td style="text-align: center;">P</td>
<td style="text-align: center;">8674</td>
</tr>
<tr class="even">
<td style="text-align: center;">sample10</td>
<td style="text-align: center;">F</td>
<td style="text-align: center;">I</td>
<td style="text-align: center;">P</td>
<td style="text-align: center;">3424</td>
</tr>
<tr class="odd">
<td style="text-align: center;">sample11</td>
<td style="text-align: center;">M</td>
<td style="text-align: center;">II</td>
<td style="text-align: center;">P</td>
<td style="text-align: center;">463</td>
</tr>
<tr class="even">
<td style="text-align: center;">sample12</td>
<td style="text-align: center;">F</td>
<td style="text-align: center;">II</td>
<td style="text-align: center;">P</td>
<td style="text-align: center;">5105</td>
</tr>
</tbody>
</table></li>
</ol>
</section>
<section id="subsetting-vectorsfactors-and-dataframes" class="level2">
<h2 class="anchored" data-anchor-id="subsetting-vectorsfactors-and-dataframes">Subsetting vectors/factors and dataframes</h2>
<ol start="2" type="1">
<li><p>Using the <code>meta</code> data frame from question #1, write out the R code you would use to perform the following operations (questions <strong>DO NOT</strong> build upon each other):</p>
<ul>
<li>return only the <code>treatment</code> and <code>sex</code> columns using <code>[]</code>:</li>
<li>return the <code>treatment</code> values for samples 5, 7, 9, and 10 using <code>[]</code>:</li>
<li>use <code>filter()</code> to return all data for those samples receiving treatment <code>P</code>:</li>
<li>use <code>filter()</code>/<code>select()</code>to return only the <code>stage</code> and <code>treatment</code> columns for those samples with <code>myc</code> &gt; 5000:</li>
<li>remove the <code>treatment</code> column from the dataset using <code>[]</code>:</li>
<li>remove samples 7, 8 and 9 from the dataset using <code>[]</code>:</li>
<li>keep only samples 1-6 using <code>[]</code>:</li>
<li>add a column called <code>pre_treatment</code> to the beginning of the dataframe with the values T, F, F, F, T, T, F, T, F, F, T, T (Hint: use <code>cbind()</code>):</li>
<li>change the names of the columns to: “A”, “B”, “C”, “D”:</li>
</ul></li>
</ol>
</section>
<section id="extracting-components-from-lists" class="level2">
<h2 class="anchored" data-anchor-id="extracting-components-from-lists">Extracting components from lists</h2>
<ol start="3" type="1">
<li><p>Create a new list, <code>list_hw</code> with three components, the <code>glengths</code> vector, the dataframe <code>df</code>, and <code>number</code> value. Use this list to answer the questions below . <code>list_hw</code> has the following structure (NOTE: the components of this list are not currently named):</p>
<pre><code>   [[1]]
   [1]   4.6  3000.0 50000.0 

   [[2]]
          species  glengths 
     1    ecoli    4.6
     2    human    3000.0
     3    corn     50000.0

   [[3]]
   [1] 8</code></pre></li>
</ol>
<p>Write out the R code you would use to perform the following operations (questions <strong>DO NOT</strong> build upon each other): - return the second component of the list: - return <code>50000.0</code> from the first component of the list: - return the value <code>human</code> from the second component: - give the components of the list the following names: “genome_lengths”, “genomes”, “record”:</p>
</section>
<section id="creating-figures-with-ggplot2" class="level2">
<h2 class="anchored" data-anchor-id="creating-figures-with-ggplot2">Creating figures with ggplot2</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="plotcounts.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">plot_image</figcaption>
</figure>
</div>
<ol start="4" type="1">
<li><p>Create the same plot as above using ggplot2 using the provided metadata and counts datasets. The <a href="https://github.com/hbc/Intro-to-R-2-day/raw/master/data/Mov10_full_meta.txt">metadata table</a> describes an experiment that you have setup for RNA-seq analysis, while the <a href="https://github.com/hbc/Intro-to-R-2-day/raw/master/data/normalized_counts.txt">associated count matrix</a> gives the normalized counts for each sample for every gene. Download the count matrix and metadata using the links provided.</p>
<p>Follow the instructions below to build your plot. Write the code you used and provide the final image.</p>
<ul>
<li><p>Read in the metadata file using: <code>meta &lt;- read.delim("Mov10_full_meta.txt", sep="\t", row.names=1)</code></p></li>
<li><p>Read in the count matrix file using: <code>data &lt;- read.delim("normalized_counts.txt", sep="\t", row.names=1)</code></p></li>
<li><p>Create a vector called <code>expression</code> that contains the normalized count values from the row in normalized_counts that corresponds to the MOV10 gene.</p></li>
<li><p>Check the class of this expression vector. Then, convert it to a numeric vector using <code>as.numeric(expression)</code></p></li>
<li><p>Bind that vector to your metadata data frame (<code>meta</code>) and call the new data frame <code>df</code>.</p></li>
<li><p>Create a ggplot by constructing the plot line by line:</p>
<ul>
<li><p>Initialize a ggplot with your <code>df</code> as input.</p></li>
<li><p>Add the <code>geom_jitter()</code> geometric object with the required aesthetics which are x and y.</p></li>
<li><p>Color the points based on <code>sampletype</code></p></li>
<li><p>Add the <code>theme_bw()</code> layer</p></li>
<li><p>Add the title “Expression of MOV10” to the plot</p></li>
<li><p>Change the x-axis label to be blank</p></li>
<li><p>Change the y-axis label to “Normalized counts”</p></li>
<li><p>Using <code>theme()</code> change the following properties of the plot:</p>
<ul>
<li><p>Remove the legend (Hint: use ?theme help and scroll down to legend.position)</p></li>
<li><p>Change the plot title size to 1.5x the default and center align</p></li>
<li><p>Change the axis title to 1.5x the default size</p></li>
<li><p>Change the size of the axis text only on the y-axis to 1.25x the default size</p></li>
<li><p>Rotate the x-axis text to 45 degrees using <code>axis.text.x=element_text(angle=45, hjust=1)</code></p></li>
</ul></li>
</ul></li>
</ul></li>
</ol>


<!-- -->

</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb2" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Introduction to R practice</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">## Creating vectors/factors and dataframes</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>We are performing RNA-Seq on cancer samples being treated with three different types of treatment (A, B, and P). You have 12 samples total, with 4 replicates per treatment. Write the R code you would use to construct your metadata table as described below.  </span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="ss">     - </span>Create the vectors/factors for each column (Hint: you can type out each vector/factor, or if you want the process go faster try exploring the <span class="in">`rep()`</span> function).</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="ss">     - </span>Put them together into a dataframe called <span class="in">`meta`</span>.</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="ss">     - </span>Use the <span class="in">`rownames()`</span> function to assign row names to the dataframe (Hint: you can type out the row names as a vector, or if you want the process go faster try exploring the <span class="in">`paste()`</span> function).</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>     Your finished metadata table should have information for the variables <span class="in">`sex`</span>, <span class="in">`stage`</span>, <span class="in">`treatment`</span>, and <span class="in">`myc`</span> levels: </span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="in">     | |sex | stage | treatment | myc |</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="in">     |:--:|:--: | :--:  | :------:  | :--: |</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="in">     |sample1|  M   |I  |A  |2343|</span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a><span class="in">     |sample2|  F   |II |A  |457|</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a><span class="in">     |sample3   |M  |II |A  |4593|</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a><span class="in">     |sample4   |F  |I  |A  |9035|</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a><span class="in">     |sample5|  M   |II |B  |3450|</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a><span class="in">     |sample6|  F|  II| B|  3524|</span></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a><span class="in">     |sample7|  M|  I|  B|  958|</span></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a><span class="in">     |sample8|  F|  II| B|  1053|</span></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a><span class="in">     |sample9|  M|  II| P|  8674|</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a><span class="in">     |sample10  |F| I   |P  |3424|</span></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a><span class="in">     |sample11| M   |II |P  |463|</span></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a><span class="in">     |sample12| F|  II| P|  5105|</span></span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a><span class="fu">## Subsetting vectors/factors and dataframes</span></span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Using the <span class="in">`meta`</span> data frame from question #1, write out the R code you would use to perform the following operations (questions **DO NOT** build upon each other):</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a><span class="ss">     - </span>return only the <span class="in">`treatment`</span> and <span class="in">`sex`</span> columns using <span class="in">`[]`</span>:</span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a><span class="ss">     - </span>return the <span class="in">`treatment`</span> values for samples 5, 7, 9, and 10 using <span class="in">`[]`</span>:</span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a><span class="ss">     - </span>use <span class="in">`filter()`</span> to return all data for those samples receiving treatment <span class="in">`P`</span>:</span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a><span class="ss">     - </span>use <span class="in">`filter()`</span>/<span class="in">`select()`</span>to return only the <span class="in">`stage`</span> and <span class="in">`treatment`</span> columns for those samples with <span class="in">`myc`</span> &gt; 5000:</span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a><span class="ss">     - </span>remove the <span class="in">`treatment`</span> column from the dataset using <span class="in">`[]`</span>:</span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true" tabindex="-1"></a><span class="ss">     - </span>remove samples 7, 8 and 9 from the dataset using <span class="in">`[]`</span>:</span>
<span id="cb2-38"><a href="#cb2-38" aria-hidden="true" tabindex="-1"></a><span class="ss">     - </span>keep only samples 1-6 using <span class="in">`[]`</span>:</span>
<span id="cb2-39"><a href="#cb2-39" aria-hidden="true" tabindex="-1"></a><span class="ss">     - </span>add a column called <span class="in">`pre_treatment`</span> to the beginning of the dataframe with the values T, F, F, F, T, T, F, T, F, F, T, T (Hint: use <span class="in">`cbind()`</span>): </span>
<span id="cb2-40"><a href="#cb2-40" aria-hidden="true" tabindex="-1"></a><span class="ss">     - </span>change the names of the columns to: "A", "B", "C", "D":</span>
<span id="cb2-41"><a href="#cb2-41" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb2-42"><a href="#cb2-42" aria-hidden="true" tabindex="-1"></a><span class="fu">## Extracting components from lists</span></span>
<span id="cb2-43"><a href="#cb2-43" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>Create a new list, <span class="in">`list_hw`</span> with three components, the <span class="in">`glengths`</span> vector, the dataframe <span class="in">`df`</span>, and <span class="in">`number`</span> value. Use this list to answer the questions below . <span class="in">`list_hw`</span> has the following structure (NOTE: the components of this list are not currently named):</span>
<span id="cb2-44"><a href="#cb2-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-45"><a href="#cb2-45" aria-hidden="true" tabindex="-1"></a><span class="in">          [[1]]</span></span>
<span id="cb2-46"><a href="#cb2-46" aria-hidden="true" tabindex="-1"></a><span class="in">          [1]   4.6  3000.0 50000.0 </span></span>
<span id="cb2-47"><a href="#cb2-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-48"><a href="#cb2-48" aria-hidden="true" tabindex="-1"></a><span class="in">          [[2]]</span></span>
<span id="cb2-49"><a href="#cb2-49" aria-hidden="true" tabindex="-1"></a><span class="in">                 species  glengths </span></span>
<span id="cb2-50"><a href="#cb2-50" aria-hidden="true" tabindex="-1"></a><span class="in">            1    ecoli    4.6</span></span>
<span id="cb2-51"><a href="#cb2-51" aria-hidden="true" tabindex="-1"></a><span class="in">            2    human    3000.0</span></span>
<span id="cb2-52"><a href="#cb2-52" aria-hidden="true" tabindex="-1"></a><span class="in">            3    corn     50000.0</span></span>
<span id="cb2-53"><a href="#cb2-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-54"><a href="#cb2-54" aria-hidden="true" tabindex="-1"></a><span class="in">          [[3]]</span></span>
<span id="cb2-55"><a href="#cb2-55" aria-hidden="true" tabindex="-1"></a><span class="in">          [1] 8</span></span>
<span id="cb2-56"><a href="#cb2-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-57"><a href="#cb2-57" aria-hidden="true" tabindex="-1"></a>Write out the R code you would use to perform the following operations (questions **DO NOT** build upon each other):</span>
<span id="cb2-58"><a href="#cb2-58" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>return the second component of the list:</span>
<span id="cb2-59"><a href="#cb2-59" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>return <span class="in">`50000.0`</span> from the first component of the list:</span>
<span id="cb2-60"><a href="#cb2-60" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>return the value <span class="in">`human`</span> from the second component: </span>
<span id="cb2-61"><a href="#cb2-61" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>give the components of the list the following names: "genome_lengths", "genomes", "record":</span>
<span id="cb2-62"><a href="#cb2-62" aria-hidden="true" tabindex="-1"></a>   </span>
<span id="cb2-63"><a href="#cb2-63" aria-hidden="true" tabindex="-1"></a><span class="fu">## Creating figures with ggplot2</span></span>
<span id="cb2-64"><a href="#cb2-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-65"><a href="#cb2-65" aria-hidden="true" tabindex="-1"></a><span class="al">![plot_image](plotcounts.png)</span></span>
<span id="cb2-66"><a href="#cb2-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-67"><a href="#cb2-67" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>Create the same plot as above using ggplot2 using the provided metadata and counts datasets. The <span class="co">[</span><span class="ot">metadata table</span><span class="co">](https://github.com/hbc/Intro-to-R-2-day/raw/master/data/Mov10_full_meta.txt)</span> describes an experiment that you have setup for RNA-seq analysis, while the <span class="co">[</span><span class="ot">associated count matrix</span><span class="co">](https://github.com/hbc/Intro-to-R-2-day/raw/master/data/normalized_counts.txt)</span> gives the normalized counts for each sample for every gene. Download the count matrix and metadata using the links provided.</span>
<span id="cb2-68"><a href="#cb2-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-69"><a href="#cb2-69" aria-hidden="true" tabindex="-1"></a>     Follow the instructions below to build your plot. Write the code you used and provide the final image.</span>
<span id="cb2-70"><a href="#cb2-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-71"><a href="#cb2-71" aria-hidden="true" tabindex="-1"></a><span class="ss">     - </span>Read in the metadata file using: <span class="in">`meta &lt;- read.delim("Mov10_full_meta.txt", sep="\t", row.names=1)`</span></span>
<span id="cb2-72"><a href="#cb2-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-73"><a href="#cb2-73" aria-hidden="true" tabindex="-1"></a><span class="ss">     - </span>Read in the count matrix file using: <span class="in">`data &lt;- read.delim("normalized_counts.txt", sep="\t", row.names=1)`</span></span>
<span id="cb2-74"><a href="#cb2-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-75"><a href="#cb2-75" aria-hidden="true" tabindex="-1"></a><span class="ss">     - </span>Create a vector called <span class="in">`expression`</span> that contains the normalized count values from the row in normalized_counts that corresponds to the MOV10 gene.  </span>
<span id="cb2-76"><a href="#cb2-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-77"><a href="#cb2-77" aria-hidden="true" tabindex="-1"></a><span class="ss">     - </span>Check the class of this expression vector. Then, convert it to a numeric vector using <span class="in">`as.numeric(expression)`</span></span>
<span id="cb2-78"><a href="#cb2-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-79"><a href="#cb2-79" aria-hidden="true" tabindex="-1"></a><span class="ss">     - </span>Bind that vector to your metadata data frame (<span class="in">`meta`</span>) and call the new data frame <span class="in">`df`</span>. </span>
<span id="cb2-80"><a href="#cb2-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-81"><a href="#cb2-81" aria-hidden="true" tabindex="-1"></a><span class="ss">     - </span>Create a ggplot by constructing the plot line by line:</span>
<span id="cb2-82"><a href="#cb2-82" aria-hidden="true" tabindex="-1"></a>     </span>
<span id="cb2-83"><a href="#cb2-83" aria-hidden="true" tabindex="-1"></a><span class="ss">          - </span>Initialize a  ggplot with your <span class="in">`df`</span> as input.</span>
<span id="cb2-84"><a href="#cb2-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-85"><a href="#cb2-85" aria-hidden="true" tabindex="-1"></a><span class="ss">          - </span>Add the <span class="in">`geom_jitter()`</span> geometric object with the required aesthetics which are x and y.</span>
<span id="cb2-86"><a href="#cb2-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-87"><a href="#cb2-87" aria-hidden="true" tabindex="-1"></a><span class="ss">          - </span>Color the points based on <span class="in">`sampletype`</span></span>
<span id="cb2-88"><a href="#cb2-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-89"><a href="#cb2-89" aria-hidden="true" tabindex="-1"></a><span class="ss">          - </span>Add the <span class="in">`theme_bw()`</span> layer </span>
<span id="cb2-90"><a href="#cb2-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-91"><a href="#cb2-91" aria-hidden="true" tabindex="-1"></a><span class="ss">          - </span>Add the title "Expression of MOV10" to the plot</span>
<span id="cb2-92"><a href="#cb2-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-93"><a href="#cb2-93" aria-hidden="true" tabindex="-1"></a><span class="ss">          - </span>Change the x-axis label to be blank</span>
<span id="cb2-94"><a href="#cb2-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-95"><a href="#cb2-95" aria-hidden="true" tabindex="-1"></a><span class="ss">          - </span>Change the y-axis label to "Normalized counts"</span>
<span id="cb2-96"><a href="#cb2-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-97"><a href="#cb2-97" aria-hidden="true" tabindex="-1"></a><span class="ss">          - </span>Using <span class="in">`theme()`</span> change the following properties of the plot:</span>
<span id="cb2-98"><a href="#cb2-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-99"><a href="#cb2-99" aria-hidden="true" tabindex="-1"></a><span class="ss">               - </span>Remove the legend (Hint: use ?theme help and scroll down to legend.position)</span>
<span id="cb2-100"><a href="#cb2-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-101"><a href="#cb2-101" aria-hidden="true" tabindex="-1"></a><span class="ss">               - </span>Change the plot title size to 1.5x the default and center align</span>
<span id="cb2-102"><a href="#cb2-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-103"><a href="#cb2-103" aria-hidden="true" tabindex="-1"></a><span class="ss">               - </span>Change the axis title to 1.5x the default size</span>
<span id="cb2-104"><a href="#cb2-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-105"><a href="#cb2-105" aria-hidden="true" tabindex="-1"></a><span class="ss">               - </span>Change the size of the axis text only on the y-axis to 1.25x the default size</span>
<span id="cb2-106"><a href="#cb2-106" aria-hidden="true" tabindex="-1"></a>               </span>
<span id="cb2-107"><a href="#cb2-107" aria-hidden="true" tabindex="-1"></a><span class="ss">               - </span>Rotate the x-axis text to 45 degrees using <span class="in">`axis.text.x=element_text(angle=45, hjust=1)`</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">This lesson has been developed by members of the teaching team at the <a href="http://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core (HBC)</a>. <br> These are open access materials distributed under the terms of the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license</a> (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited. <br> Some materials used are adapted from work that is Copyright © <a href="http://datacarpentry.org/">Data Carpentry</a>. All Data Carpentry instructional material is made available under the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license</a> (CC BY 4.0).</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>



</body></html>